// leanft.report.builder/lib/logger.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates.
/* Created on 06/10/2015. */

/**
 * Methods for writing log messages at different levels.
 * @type {{trace: Function, debug: Function, info:Function, warn:Function, error:Function, fatal:Function}}
 * @class
 * @static
 */
var Logger = function(category,level){
    this.category = category;
    this.level = level || Logger.prototype.DEFAULT_LEVEL;
    this.appender = new ConsoleAppender();
};


Logger.prototype = {
    category: null,
    level: 0,
    DEFAULT_LEVEL: 0, //off
    appender: null,

    logLevelAsString: {
        1000: "Fatal",
        2000: "Error",
        3000: "Warn",
        4000: "Info",
        5000: "Debug",
        6000: "Trace"
    },

	/**
	* Writes a log entry.
	* @function
	* @alias trace
	* @memberof! Logger
	*/
    trace: function (msg) {
        this.writeToLog(msg,6000);
    },

    /**
	* Writes a log entry.
	* @function
	* @memberof! Logger
	*/
	debug: function (msg) {
        this.writeToLog(msg,5000);
    },

     /**
	* Writes a log entry.
	* @function
	* @memberof! Logger
	*/
   info: function (msg) {
        this.writeToLog(msg,4000);
    },

     /**
	* Writes a log entry.
	* @function
	* @memberof! Logger
	*/
   warn: function (msg) {
        this.writeToLog(msg,3000);
    },

    /**
	* Writes a log entry.
	* @function
	* @memberof! Logger
	*/
    error: function (msg) {
        this.writeToLog(msg,2000);
    },

     /**
	* Writes a log entry.
	* @function
	* @memberof! Logger
	*/
   fatal: function (msg) {
        this.writeToLog(msg,1000);
    },

    writeToLog: function (msg, level) {
        if(this.level <= level)
            return;

        this.appender.append(this.formatMessage(msg,level),Logger.prototype.logLevelAsString[level]);
    },

    formatMessage: function(msg,level){
        return new Date().toGMTString() + " [" + Logger.prototype.logLevelAsString[level] + "] " + this.category + ": " + msg;
    }
};

function ConsoleAppender(){

}

ConsoleAppender.prototype = {
    appenderFunctionByLevel: {
        "trace": console.log,
        "debug": console.log,
        "info": console.info,
        "warn": console.warn,
        "error": console.error,
        "fatal": console.error
    },
    append: function(msg,levelString){
        this.appenderFunctionByLevel[levelString.toLowerCase()].call(console,msg);
    }
};

module.exports.Logger = Logger;
