// leanft.sdk.oracle/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var OracleFactory = require("./lib/oracle_factory.js").OracleFactory;
var DescriptionsAndEnums = {
    ApplicationSession: require("./lib/application_session.js").ApplicationSession,
    FlexFieldWindow: require("./lib/flex_field_window.js").FlexFieldWindow,
    FormWindow: require("./lib/form_window.js").FormWindow,
    Button: require("./lib/button.js").Button,
    LogonWindow: require("./lib/logon_window.js").LogonWindow,
    NavigatorWindow: require("./lib/navigator_window.js").NavigatorWindow,
    PickList: require("./lib/pick_list.js").PickList,
    Calendar: require("./lib/calendar.js").Calendar,
    CheckBox: require("./lib/check_box.js").CheckBox,
    StatusBar: require("./lib/status_bar.js").StatusBar,
    Notification: require("./lib/notification.js").Notification,
    List: require("./lib/list.js").List,
    ListOfValues: require("./lib/list_of_values.js").ListOfValues,
    RadioGroup: require("./lib/radio_group.js").RadioGroup,
    Tab: require("./lib/tab.js").Tab,
    EditField: require("./lib/edit_field.js").EditField,
    TreeView: require("./lib/tree_view.js").TreeView,
    Table: require("./lib/table.js").Table,
    SoftKeys: require("./lib/oracle_sendkey_behavior.js").SoftKeys,
    NotificationType: require("./lib/notification.js").NotificationType
};

module.exports.OracleFactory = OracleFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
