// leanft.sdk.uiapro/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var UIAProFactory = require("./lib/uiapro_factory.js").UIAProFactory;
var DescriptionsAndEnums = {
    Window: require("./lib/window.js").Window,
    UiObject: require("./lib/ui_object.js").UiObject,
    Button: require("./lib/button.js").Button,
    Calendar: require("./lib/calendar.js").Calendar,
    CheckBox: require("./lib/check_box.js").CheckBox,
    ComboBox: require("./lib/combo_box.js").ComboBox,
    Edit: require("./lib/edit.js").Edit,
    Link: require("./lib/link.js").Link,
    SplitButton: require("./lib/split_button.js").SplitButton,
    RadioButton: require("./lib/radio_button.js").RadioButton,
    Menu: require("./lib/menu.js").Menu,
    List: require("./lib/list.js").List,
    ListItem: require("./lib/list_item.js").ListItem,
    Slider: require("./lib/slider.js").Slider,
    TreeView: require("./lib/tree_view.js").TreeView,
    TabControl: require("./lib/tab_control.js").TabControl,
    Table: require("./lib/table.js").Table,
    AppBar: require("./lib/app_bar.js").AppBar,
    Document: require("./lib/document.js").Document,
    GridItem: require("./lib/grid_item.js").GridItem,
    Grid: require("./lib/grid.js").Grid,
    Group: require("./lib/group.js").Group,
    HeaderItem: require("./lib/header_item.js").HeaderItem,
    Header: require("./lib/header.js").Header,
    Image: require("./lib/image.js").Image,
    MenuBar: require("./lib/menu_bar.js").MenuBar,
    MenuItem: require("./lib/menu_item.js").MenuItem,
    Pane: require("./lib/pane.js").Pane,
    ProgressBar: require("./lib/progress_bar.js").ProgressBar,
    ScrollBar: require("./lib/scroll_bar").ScrollBar,
    SemanticZoom: require("./lib/semantic_zoom.js").SemanticZoom,
    Separator: require("./lib/separator.js").Separator,
    Spinner: require("./lib/spinner.js").Spinner,
    StatusBar: require("./lib/status_bar.js").StatusBar,
    TabItem: require("./lib/tab_item.js").TabItem,
    Text: require("./lib/text.js").Text,
    Thumb: require("./lib/thumb.js").Thumb,
    TitleBar: require("./lib/title_bar.js").TitleBar,
    ToolBar: require("./lib/tool_bar.js").ToolBar,
    ToolTip: require("./lib/tool_tip.js").ToolTip,
    TreeViewItem: require("./lib/tree_view_item").TreeViewItem,
    PatternType: require("./lib/ui_object_base.js").PatternType,
    ScrollAmount: require("./lib/patterns/scroll_pattern.js").ScrollAmount,
    TableTraversalDirection: require("./lib/patterns/table_pattern.js").TableTraversalDirection,
    SupportedTextSelection: require("./lib/patterns/text_pattern.js").SupportedTextSelection,
    TextUnit: require("./lib/patterns/text_pattern.js").TextUnit,
    Direction: require("./lib/patterns/text_pattern.js").Direction,
    Alignment: require("./lib/patterns/text_pattern.js").Alignment,
    SelectOptions: require("./lib/patterns/legacy_accessible_pattern.js").SelectOptions,
    AccessibleRole: require("./lib/patterns/legacy_accessible_pattern.js").AccessibleRole,
    AccessibleState: require("./lib/patterns/legacy_accessible_pattern.js").AccessibleState,
    WindowInteractionState: require("./lib/patterns/window_pattern.js").WindowInteractionState,
    WindowVisualState: require("./lib/patterns/window_pattern.js").WindowVisualState,
    ExpandCollapseState: require("./lib/patterns/expand_collapse_pattern.js").ExpandCollapseState
};

module.exports.UIAProFactory = UIAProFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
