// leanft.sdk.stdwin/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var StdWindowFactory = require("./lib/stdwindow_factory.js").StdWindowFactory;
var DescriptionsAndEnums = {
    Window: require("./lib/window.js").Window,
    Dialog: require("./lib/dialog.js").Dialog,
    Button: require("./lib/button.js").Button,
    UiObject: require("./lib/ui_object.js").UiObject,
    Edit: require("./lib/edit.js").Edit,
    ListBox: require("./lib/list_box.js").ListBox,
    ListView: require("./lib/list_view.js").ListView,
    TreeView: require("./lib/tree_view.js").TreeView,
    ToolBar: require("./lib/tool_bar.js").ToolBar,
    Menu: require("./lib/menu.js").Menu,
    MenuType: require("./lib/menu.js").MenuType,
    CheckBox: require("./lib/check_box.js").CheckBox,
    RadioButton: require("./lib/radio_button.js").RadioButton,
    StatusBar: require("./lib/status_bar.js").StatusBar,
    TabControl: require("./lib/tab_control.js").TabControl,
    ComboBox: require("./lib/combo_box.js").ComboBox,
    Editor: require("./lib/editor.js").Editor,
    ScrollBar: require("./lib/scroll_bar.js").ScrollBar,
    SpinBox: require("./lib/spin_box.js").SpinBox,
    Static: require("./lib/static.js").Static,
    Calendar: require("./lib/calendar.js").Calendar
};

//Internal Behaviors to be used by PowerBuilder
module.exports.Behaviors = require("./lib/std_behaviors.js");

module.exports.StdWindowFactory = StdWindowFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
