// leanft.sdk.stdwin/lib/check_box.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/17/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var CheckedState = Core.CheckedState;
var UiObjectBase = require("./ui_object_base.js");
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var UiWithLayoutInfo = require("./ui_object_base_with_layout_info.js");
var UiObjectWithLayoutInfoBehaviour = UiWithLayoutInfo.UiObjectWithLayoutInfoBehaviour;
var UiObjectWithLayoutInfoDescription = UiWithLayoutInfo.UiObjectWithLayoutInfoDescription;
var CheckedStateDescriptionBehavior = require("./checked_behavior.js").CheckedStateDescriptionBehavior;
var CheckedStateTOBehavior = require("./checked_behavior.js").CheckedStateTOBehavior;

/**
 * CheckBoxDescriptionBehavior - the description for a Checkbox object
 * @memberof StdWin
 * @ignore
 */
var CheckBoxDescriptionBehavior = {
    /**
     * The state of the checkbox.
     * @param {CheckedState} [state] When provided, this method acts as a setter and sets the state value.
     * @returns {CheckedState | Description} If no parameter is provided, the method acts as a getter and returns the current stored state value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    state: function (state) {
        return this._state(state);
    }
};

/**
 * The description of a CheckBox test object.
 * @param {Object} [props] Initializes a CheckBox test object using the supplied property bag.
 * @returns {CheckBox} the description object that was initialized.
 * @class CheckBox
 * @memberof StdWin
 * @extends StdWin.UiObjectBaseDescription
 * @mixes StdWin.UiObjectWithLayoutInfoDescription
 * @mixes StdWin.CheckBoxDescriptionBehavior
 */
function CheckBox(props) {
    var checkBoxDesc = new UiObjectBaseDescription("CheckBox");
    Utils.extend(checkBoxDesc, UiObjectWithLayoutInfoDescription);
    Utils.extend(checkBoxDesc, CheckedStateDescriptionBehavior);
    Utils.extend(checkBoxDesc, CheckBoxDescriptionBehavior);
    checkBoxDesc._defineByPropertyBag(props);
    return checkBoxDesc;
}

/**
 * A CheckBox object
 * @constructor
 * @class CheckBoxTO
 * @extends StdWin.UiObjectBaseTO
 * @mixes StdWin.UiObjectWithLayoutInfoBehaviour._properties
 * @mixes StdWin.CheckBoxTOBehavior._methods
 * @mixes StdWin.CheckBoxTOBehavior._properties
 * @memberof StdWin
 */
function CheckBoxTO(session, description){
    description = description || CheckBox();
    var checkBoxTO = new UiObjectBaseTO("WinCheckBox", session, description);
    checkBoxTO._mergeBehavior(UiObjectWithLayoutInfoBehaviour);
    checkBoxTO._mergeBehavior(CheckedStateTOBehavior);
    checkBoxTO._mergeBehavior(CheckBoxTOBehavior);
    return checkBoxTO;
}

/**
 * The behavior of the CheckBox object.
 * @memberof StdWin
 * @ignore
 */
var CheckBoxTOBehavior = {
    /**
     * @ignore
     * @memberof StdWin.CheckBoxTOBehavior
     */
    _properties: {
        /**
         * Gets the state of the checkbox.
         * @function
         * @returns {Promise<CheckedState>} a promise that is fulfilled with the state of the checkbox.
         */
        state: {
            type: 'enum',
            enumType: Core.CheckedState,
            enumName: 'CheckedState',
            wiredName: CheckedStateTOBehavior._checkedWiredName,
            getter: function () {
                return this._state();
            }
        }
    },

    /**
     * @ignore
     * @memberof StdWin.CheckBoxTOBehavior
     */
    _methods: {
        /*jshint eqnull:true */
        /**
         * Sets the state of checkbox.
         * @function
         * @param {CheckedState|string} state The status to be assigned to the checkbox.
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        setState: function (state) {
            if (state == null)
                throw new Error("The parameter state cannot be null.");
            return this._executeWithEvents(arguments, "Set", CheckedState._checkedStateToNative(state));
        }
    }
};

module.exports.CheckBox = CheckBox;
module.exports.CheckBoxTO = CheckBoxTO;
module.exports.CheckBoxDescriptionBehavior = CheckBoxDescriptionBehavior;
module.exports.CheckBoxTOBehavior = CheckBoxTOBehavior;
