// leanft.sdk.stdwin/lib/combo_box.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/17/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var StringUtils = Core.StringUtils;
var UiObjectBase = require("./ui_object_base.js");
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var UiWithLayoutInfo = require("./ui_object_base_with_layout_info.js");
var UiObjectWithLayoutInfoBehaviour = UiWithLayoutInfo.UiObjectWithLayoutInfoBehaviour;
var UiObjectWithLayoutInfoDescription = UiWithLayoutInfo.UiObjectWithLayoutInfoDescription;
var Convert = Core.Convert;

/**
 * ComboBoxDescriptionBehavior - the description for a ComboBox object
 * @memberof StdWin
 * @ignore
 */
var ComboBoxDescriptionBehavior = {
    /**
     * An array of all items in the combo box.
     * @param {string[]} [items] When provided, this method acts as a setter and sets the items value.
     * @returns {string[] | Description} If no parameter is provided, the method acts as a getter and returns the current stored items value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    items: function (items) {
        return this._propertyHandler("all items", items, StringUtils.joinFunc(";"), StringUtils.splitFunc(";"));
    },

    /**
     * The selected items in the combo box.
     * @param {string} [selectedItem] When provided, this method acts as a setter and sets the selectedItems value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored selectedItems value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    selectedItem: function (selectedItem) {
        return this._simplePropertyHandler("selection", selectedItem);
    },

    /**
     * The number of items in the combo box.
     * @param {number} [itemsCount] When provided, this method acts as a setter and sets the itemsCount value.
     * @returns {number | Description} If no parameter is provided, the method acts as a getter and returns the current stored itemsCount value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    itemsCount: function (itemsCount) {
        return this._propertyHandler("items count", itemsCount, Convert.toNumber);
    }
};

/**
 * The description of a ComboBox test object.
 * @param {Object} [props] Initializes a ComboBox test object using the supplied property bag.
 * @returns {ComboBox} the description object that was initialized.
 * @class ComboBox
 * @memberof StdWin
 * @extends StdWin.UiObjectBaseDescription
 * @mixes StdWin.UiObjectWithLayoutInfoDescription
 * @mixes StdWin.ComboBoxDescriptionBehavior
 */
function ComboBox(props) {
    var comboBoxDesc = new UiObjectBaseDescription("ComboBox");
    Utils.extend(comboBoxDesc, UiObjectWithLayoutInfoDescription);
    Utils.extend(comboBoxDesc, ComboBoxDescriptionBehavior);
    comboBoxDesc._defineByPropertyBag(props);
    return comboBoxDesc;
}

/**
 * A ComboBox object
 * @constructor
 * @class ComboBoxTO
 * @extends StdWin.UiObjectBaseTO
 * @mixes StdWin.UiObjectWithLayoutInfoBehaviour._properties
 * @mixes StdWin.ComboBoxTOBehavior._properties
 * @mixes StdWin.ComboBoxTOBehavior._methods
 * @mixes StdWin.ComboBoxTOOpenMethodBehavior._methods
 * @memberof StdWin
 */
function ComboBoxTO(session, description) {
    description = description || ComboBox();
    var comboBoxTO = new UiObjectBaseTO("WinComboBox", session, description);
    comboBoxTO._mergeBehavior(UiObjectWithLayoutInfoBehaviour);
    comboBoxTO._mergeBehavior(ComboBoxTOBehavior);
    comboBoxTO._mergeBehavior(ComboBoxTOOpenMethodBehavior);
    return comboBoxTO;
}

//Consts
var allItemsWiredName = "all items";

/**
 * The behavior of the ComboBox open method behavior.
 * @memberof StdWin
 * @ignore
 */
var ComboBoxTOOpenMethodBehavior = {
    /**
     * @ignore
     * @memberof StdWin.ComboBoxTOOpenMethodBehavior
     */
    _methods: {
        /**
         * Opens the combo box.
         * @function
         * @returns {Promise<void>} a promise that is fulfilled when the operation ends.
         */
        open: function () {
            return this._executeWithEvents(arguments, "Open");
        }
    }
};

/**
 * The behavior of the ComboBox object.
 * @memberof StdWin
 * @ignore
 */
var ComboBoxTOBehavior = {
    /**
     * @ignore
     * @memberof StdWin.ComboBoxTOBehavior
     */
    _properties: {
        /**
         * Gets the selected item of the combo box.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the selected item of the combo box
         */
        selectedItem: {
            wiredName: "selection"
        },
        /**
         * Gets all items of the combo box.
         * @function
         * @returns {Promise<string[]>} a promise that is fulfilled with all items of the combo box.
         */
        items: {
            wiredName: allItemsWiredName,
            getter: function () {
                return this._getROProperty(allItemsWiredName, arguments).then(
                    StringUtils.splitFunc("\n")
                );
            }
        }
    },

    /**
     * @ignore
     * @memberof StdWin.ComboBoxTOBehavior
     */
    _methods: {
        /*jshint eqnull:true */
        /**
         * Selects an item in the combo box.
         * @function
         * @param {number|string} indexOrItem The item to be selected.
         * The input can be the 0-based index of the item or the text of the item.
         * @returns {Promise<void>} a promise that is fulfilled when the select operation ends.
         */
        select: function (indexOrItem) {
            if (indexOrItem == null)
                throw new Error("The parameter indexOrItem cannot be null.");
            return this._executeWithEvents(arguments, "Select", indexOrItem);
        }
    },

    _additionalProperties: {
        itemsCount: {wiredName: "items count"}
    }
};

module.exports.ComboBox = ComboBox;
module.exports.ComboBoxTO = ComboBoxTO;
module.exports.ComboBoxDescriptionBehavior = ComboBoxDescriptionBehavior;
module.exports.ComboBoxTOBehavior = ComboBoxTOBehavior;
