// leanft.sdk.stdwin/lib/list_view.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/13/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var UiObjectBase = require("./ui_object_base.js");
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var UiWithLayoutInfo = require("./ui_object_base_with_layout_info.js");
var UiObjectWithLayoutInfoBehaviour = UiWithLayoutInfo.UiObjectWithLayoutInfoBehaviour;
var UiObjectWithLayoutInfoDescription = UiWithLayoutInfo.UiObjectWithLayoutInfoDescription;
var ListBaseBehavior = Core.ListBaseBehavior;
var ListViewItem = require("./listview_item.js").ListViewItem;
var ListBoxDescriptionBaseBehavior = Core.ListBoxDescriptionBaseBehavior;
var ContainerBehavior = Core.ContainerBehavior;
var WrappedItemExecutorBehavior = Core.ExecuteItemsWithEventsBehavior;
var ListViewTOBaseBehavior = Core.ListViewTOBaseBehavior;

/**
 * The description of a ListView test object.
 * @param {Object} [props] Initializes a ListView test object using the supplied property bag.
 * @returns {ListView} the description object that was initialized.
 * @class ListView
 * @memberof StdWin
 * @extends StdWin.UiObjectBaseDescription
 * @mixes StdWin.UiObjectWithLayoutInfoDescription
 * @mixes ListBoxDescriptionBaseBehavior
 */
function ListView(props) {
    var listViewDesc = new UiObjectBaseDescription("ListView");
    Utils.extend(listViewDesc, UiObjectWithLayoutInfoDescription);
    Utils.extend(listViewDesc, ListBoxDescriptionBaseBehavior);
    listViewDesc._defineByPropertyBag(props);
    return listViewDesc;
}

/**
 * A ListView object
 * @constructor
 * @class ListViewTO
 * @extends StdWin.UiObjectBaseTO
 * @mixes StdWin.UiObjectWithLayoutInfoBehaviour._properties
 * @mixes ListBaseBehavior._methods
 * @mixes ListViewTOBaseBehavior._methods
 * @mixes ListViewTOBaseBehavior._properties
 * @mixes StdWin.ListViewTOBehavior._properties
 * @mixes StdWin.ListViewTOBehavior._methods
 * @memberof StdWin
 */
function ListViewTO(session, description) {
    description = description || ListView();
    var listViewTO = new UiObjectBaseTO("WinListView", session, description);
    listViewTO._mergeBehavior(UiObjectWithLayoutInfoBehaviour);
    listViewTO._mergeBehavior(ContainerBehavior);
    listViewTO._mergeBehavior(WrappedItemExecutorBehavior);
    listViewTO._mergeBehavior(ListBaseBehavior);
    listViewTO._mergeBehavior(ListViewTOBaseBehavior);
    listViewTO._mergeBehavior(ListViewTOBehavior);
    return listViewTO;
}

/**
 * The behavior of the ListView object
 * @memberof StdWin
 * @ignore
 */
var ListViewTOBehavior = {
    /**
     * @ignore
     * @memberof StdWin.ListViewTOBehavior
     */
    _properties: {
        /**
         * The collection of items in the list as a read-only collection object.
         * @function
         * @returns {Promise<ListViewItem[]>} a promise that is fulfilled with collection of all list items in the list view. See {@link StdWin.ListViewItem}
         */
        items: {
            wiredName: ContainerBehavior._allItemsDefaultWiredName,
            getter: function () {
                return this._items("\n");
            }
        },

        /**
         * A list of all checked items as a read-only collection object.
         * @function
         * @returns {Promise<ListViewItem[]>} a promise that is fulfilled with a list of all checked item in the list view. See {@link StdWin.ListViewItem}
         */
        selectedItems: {
            wiredName: ContainerBehavior._selectedItemsDefaultWiredName,
            getter: function () {
                return this._selectedItems("\n");
            }
        },

        /**
         * A list of all checked items as a read-only collection object.
         * @function
         * @returns {Promise<ListViewItem[]>} a promise that is fulfilled with a list of all checked items in the list view. See {@link StdWin.ListViewItem}
         */
        checkedItems: {
            wiredName: ContainerBehavior._checkedItemsDefaultWiredName,
            getter: function () {
                return this._checkedItems(";", "\n");
            }
        }
    },

    /**
     * @ignore
     * @memberof StdWin.ListViewTOBehavior
     */
    _methods: {
        /**
         * Activates (double-clicks) the specified item in the list.
         * @function
         * @param {ListViewItem|number|string} item The specified item to activate from the list.
         * The input can be the list item, the 0-based index of the item, or the name/text of the item.  See {@link StdWin.ListViewItem}
         * @param {MouseButton} [button = MouseButton.left] The button to activate. Default: the left button.
         * @returns {Promise<void>} a promise that is fulfilled when the activate operation is performed.
         */
        activateItem: function (item, button) {
            return this._activateItem(item, button);
        },

        /**
         * Deselects a list of items.
         * @function
         * @param {...(ListViewItem | number | string)} items You can pass as many items as you like. Each item can be either the ListViewItem, the name of the item you wish to select,
         * or the numeric index of the item in the list. See {@link StdWin.ListViewItem}
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        deselect: function (items) {
            return this._deselect.apply(this, arguments);
        },

        /**
         * Selects a list of items.
         * @function
         * @param {...(ListViewItem | number | string)} items You can pass as many items as you like. Each item can be either the ListViewItem, the name of the item you wish to select,
         * or the numeric index of the item in the list. See {@link StdWin.ListViewItem}
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        select: function (items) {
            return this._select.apply(this, arguments);
        },

        /**
         * Selects additional items from the multi-selection list.
         * <BR>Remark: This method is applicable only to lists that allow multiple selection.
         * @function
         * @param {...(ListViewItem | string | number)} items One or more items to add to the selection. Each item can be either a {@link StdWin.ListViewItem}, the item's name,
         * or the item's numeric index in the list.
         * @returns {Promise<void>} A promise that is fulfilled when the operation is performed.
         */
        extendSelect: function (items) {
            return this._extendSelect.apply(this, arguments);
        },

        /**
         * Returns an item with the given text
         * @function
         * @param {string} text The text of the specified item.
         * @returns {ListViewItem} The ListViewItem object. See {@link StdWin.ListViewItem}
         */
        getItem: function (text) {
            return this._getItem(text);
        }
    },

    _helpers: {
        /**
         * Creates an item from the identifier and text representation.
         * @param {number|string} identifier The identifier of the item. If it is a number, it is the 0-based index of the item.
         * @param {string} textRepresentation The text representation of the item.
         * @returns {ListViewItem} The ListViewItem object. See {@link StdWin.ListViewItem}
         * @ignore
         */
        _createItem: function (identifier, textRepresentation) {
            return new ListViewItem(this, identifier, textRepresentation);
        }
    },

    _additionalProperties: {
        itemsCount: {wiredName: ContainerBehavior._itemsCountDefaultWiredName}
    }
};

module.exports.ListView = ListView;
module.exports.ListViewTO = ListViewTO;
module.exports.ListViewTOBehavior = ListViewTOBehavior;
module.exports.ListViewItem = ListViewItem;
