// leanft.sdk.stdwin/lib/menu_item.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/16/2016.
 */
var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var MergeBehavior = Core.MergeBehavior;
/**
 * A MenuItem object
 * @constructor
 * @class MenuItem
 * @mixes StdWin.MenuItemBehavior._methods
 * @mixes StdWin.MenuItemBehavior._properties
 * @memberof StdWin
 */
function MenuItem(ownerMenu, itemPath) {
    Utils.extend(this, MergeBehavior);
    this._mergeBehavior(MenuItemBehavior);
    this._ownerMenu = ownerMenu;
    this._path = itemPath;
    this._executeWithEventsOnItemOnError = this._ownerMenu._executeWithEventsOnItemOnError.bind(this._ownerMenu, this);
}

/**
 * The methods of MenuItem object
 * @memberof StdWin
 * @ignore
 */
MenuItem.prototype = {
    _ownerMenu: null,
    _path: null,
    _executeWithEventsOnItemOnError: null
};

var MenuItemBehavior = {
    /**
     * @memberof StdWin.MenuItemBehavior
     * @ignore
     */
    _properties: {
        /**
         * Gets the path of the menu item.
         * @returns {string} The path of the menu item.
         * @ignore
         */
        path: function () {
            return this._path;
        },

        /**
         * Checks if the menu item is checked.
         * @function
         * @returns {Promise<boolean>} a promise that is fulfilled with an indication of whether the menu item is checked or not.
         */
        isChecked: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._isItemChecked, this._path);
        },

        /**
         * Checks if the menu item is enabled.
         * @function
         * @returns {Promise<boolean>} a promise that is fulfilled with an indication of whether the menu item is enabled or not.
         */
        isEnabled: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._isItemEnabled, this._path);
        },

        /**
         * Checks if the menu item exists.
         * @function
         * @returns {Promise<boolean>} a promise that is fulfilled with an indication of whether the menu item exists or not.
         */
        exists: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._isItemExists, this._path);
        },

        /**
         * Checks if the menu item has sub-menus.
         * @function
         * @returns {Promise<boolean>} a promise that is fulfilled with an indication of whether the menu item has sub-menus or not.
         */
        hasSubMenu: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._itemHasSubMenu, this._path);
        },

        /**
         * Gets the index of the menu item.
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the index of the menu item.
         */
        index: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._getItemIndex, this._path);
        },

        /**
         * Gets the label of the menu item.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the label of the menu item.
         */
        label: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._getItemLabel, this._path);
        },

        /**
         * Gets the number of sub-menus of the menu item.
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the number of sub-menus of the menu item.
         */
        subMenuCount: function () {
            return this._executeWithEventsOnItemOnError(this._ownerMenu._getItemSubMenuCount, this._path);
        }
    }
};

module.exports.MenuItem = MenuItem;
