// leanft.sdk.stdwin/lib/std_behaviors.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 3/19/2018.
 */

// Descriptions
module.exports.UiObjectBaseDescriptionBehavior = require("./ui_object_base.js").UiObjectBaseDescriptionBehavior;
module.exports.UiObjectWithLayoutInfoDescription = require("./ui_object_base_with_layout_info.js").UiObjectWithLayoutInfoDescription;
module.exports.CheckBoxDescriptionBehavior = require("./check_box.js").CheckBoxDescriptionBehavior;
module.exports.CheckedStateDescriptionBehavior = require("./checked_behavior.js").CheckedStateDescriptionBehavior;
module.exports.RadioButtonDescriptionBehavior = require("./radio_button.js").RadioButtonDescriptionBehavior;
module.exports.ComboBoxDescriptionBehavior = require("./combo_box.js").ComboBoxDescriptionBehavior;
module.exports.ScrollBarDescriptionBehavior = require("./scroll_bar.js").ScrollBarDescriptionBehavior;
module.exports.WindowDescriptionBehavior = require("./window.js").WindowDescriptionBehavior;

// Test Objects
module.exports.UiObjectBaseTOBehavior = require("./ui_object_base.js").UiObjectBaseTOBehavior;
module.exports.UiObjectWithLayoutInfoBehaviour = require("./ui_object_base_with_layout_info.js").UiObjectWithLayoutInfoBehaviour;
module.exports.CheckBoxTOBehavior = require("./check_box.js").CheckBoxTOBehavior;
module.exports.CheckedStateTOBehavior= require("./checked_behavior.js").CheckedStateTOBehavior;
module.exports.RadioButtonTOBehavior = require("./radio_button.js").RadioButtonTOBehavior;
module.exports.ComboBoxTOBehavior = require("./combo_box.js").ComboBoxTOBehavior;
module.exports.EditTOBehavior = require("./edit.js").EditTOBehavior;
module.exports.ListBoxTOBehavior = require("./list_box.js").ListBoxTOBehavior;
module.exports.ListItem = require("./list_box.js").ListItem;
module.exports.ListViewTOBehavior = require("./list_view.js").ListViewTOBehavior;
module.exports.ListViewItem = require("./list_view.js").ListViewItem;
module.exports.ScrollBarTOBehavior = require("./scroll_bar.js").ScrollBarTOBehavior;
module.exports.TabControlTOBehavior = require("./tab_control.js").TabControlTOBehavior;
module.exports.Tab = require("./tab_control.js").Tab;
module.exports.ToolBarTOBehavior = require("./tool_bar.js").ToolBarTOBehavior;
module.exports.ToolBarButton = require("./tool_bar.js").ToolBarButton;
module.exports.TreeViewTOBehavior = require("./tree_view.js").TreeViewTOBehavior;
module.exports.TreeViewNode = require("./tree_view.js").TreeViewNode;
module.exports.WindowTOBehavior = require("./window.js").WindowTOBehavior;
