// leanft.report/mocha/mochaReporter.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..

function MochaReporter(runner) {
  var reporter = null;
  if(global.__LFT_REPORTER__){
    reporter = global.__LFT_REPORTER__;
    this.onSuiteStart = reporter.onSuiteStart.bind(reporter);
    this.onSuiteDone = reporter.onSuiteDone.bind(reporter);
    this.onSpecStart = reporter.onSpecStart.bind(reporter);
    this.onSpecDone = reporter.onSpecDone.bind(reporter);
    this.onStart = reporter.onStart.bind(reporter);
    this.onDone = reporter.onDone.bind(reporter);
  }

  reporter = this;

  runner.on('start', function(){
    reporter.fireEvent("onStart");
  });

  runner.on('end', function(){
    reporter.fireEvent("onDone");
  });

  runner.on('suite', function(suite){
    if (!suite.title)
      return;

    reporter.fireEvent("onSuiteStart", {name: suite.title});
  });

  runner.on('suite end', function(suite){
    reporter.fireEvent("onSuiteDone", {name: suite.title, status: suite.state});
  });

  runner.on('test', function(test){
    reporter.fireEvent("onSpecStart", {name: test.title});
  });

  runner.on('test end', function (test) {
    if (test.state === 'failed') {
      //already handled by the on fail callback
        return;
    }
    reporter.fireEvent("onSpecDone", {name: test.title, status: test.state});
  });

  runner.on('fail', function (test, err) {
    reporter.fireEvent("onSpecDone", {name: test.title, status: test.state, error: err});
    console.error(test.fullTitle(), "-- Failure: ", err.stack);
  });
}

MochaReporter.prototype = {
  onSuiteStart: null,
  onSuiteDone: null,
  onSpecStart: null,
  onSpecDone: null,
  onDone: null,
  onStart:null,

  fireEvent: function(eventName, data){
    if(!this[eventName]){
      return;
    }

    this[eventName].call(undefined, data);
  }
};

module.exports = MochaReporter;
