// leanft.sdk.wpf/lib/combo_box.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 05/24/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var UiObjectBase = require("./ui_object_base.js");
var StringUtils = Core.StringUtils;
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var Convert = Core.Convert;

/**
 * ComboBoxDescriptionBehavior - the description for a ComboBox object
 * @memberof WPF
 * @ignore
 */
var ComboBoxDescriptionBehavior = {
    /**
     * The help text associated with the element.
     * @param {string} [helpText] When provided, this method acts as a setter and sets the helpText value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored helpText value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    helpText: function (helpText) {
        return this._simplePropertyHandler("helptext", helpText);
    },

    /**
     * The name of the object.
     * @param {string} [name] When provided, this method acts as a setter and sets the name value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored name value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    name: function (name) {
        return this._simplePropertyHandler("name", name);
    },

    /**
     * The text associated with the parent object of the object.
     * @param {string} [parentText] When provided, this method acts as a setter and sets the parentText value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored parentText value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    parentText: function (parentText) {
        return this._simplePropertyHandler("parent text", parentText);
    },

    /**
     * An array of all items in the combobox.
     * @param {string[]} [items] When provided, this method acts as a setter and sets the items value.
     * @returns {string[] | Description} If no parameter is provided, the method acts as a getter and returns the current stored items value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    items: function (items) {
        return this._propertyHandler("all items", items, StringUtils.joinFunc("\n"), StringUtils.splitFunc("\n"));
    },

    /**
     * The selected items in the combobox.
     * @param {string} [selectedItem] When provided, this method acts as a setter and sets the selectedItems value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored selectedItems value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    selectedItem: function (selectedItem) {
        return this._simplePropertyHandler("selection", selectedItem);
    }
};

/**
 * The description of a WPF ComboBox test object.
 * @param {Object} [props] Initializes a ComboBox test object using the supplied property bag.
 * @returns {ComboBox} the description object that was initialized.
 * @class ComboBox
 * @extends WPF.UiObjectBaseDescription
 * @mixes WPF.ComboBoxDescriptionBehavior
 * @memberof WPF
 */
function ComboBox(props) {
    var comboBoxDesc = new UiObjectBaseDescription("ComboBox");
    Utils.extend(comboBoxDesc, ComboBoxDescriptionBehavior);
    comboBoxDesc._defineByPropertyBag(props);
    return comboBoxDesc;
}

/**
 * A WPF ComboBox object
 * @constructor
 * @class ComboBoxTO
 * @extends WPF.UiObjectBaseTO
 * @mixes WPF.ComboBoxTOBehavior._properties
 * @mixes WPF.ComboBoxTOBehavior._methods
 * @memberof WPF
 */
function ComboBoxTO(session, description) {
    description = description || ComboBox();
    var comboBoxTO = new UiObjectBaseTO("WpfComboBox", session, description);
    comboBoxTO._mergeBehavior(ComboBoxTOBehavior);
    return comboBoxTO;
}

//Consts
var allItemsWiredName = "all items";

/**
 * The behavior of ComboBox object
 * @memberof WPF
 * @ignore
 */
var ComboBoxTOBehavior = {
    /**
     * @ignore
     * @memberof WPF.ComboBoxTOBehavior
     */
    _properties: {
        /**
         * The help text associated with the object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the combobox's help text.
         */
        helpText: {
            wiredName: "helptext"
        },

        /**
         * The name of the object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the combobox's name.
         */
        name: {

        },

        /**
         * The text associated with the parent object of the object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the comboBox's parent text.
         */
        parentText: {
            wiredName: "parent text"
        },

        /**
         * Gets the selected item of the combobox.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the selected item of the combobox.
         */
        selectedItem: {
            wiredName: "selection"
        },

        /**
         * Gets all items of the combobox.
         * @function
         * @returns {Promise<string[]>} a promise that is fulfilled with the all the items of the combobox.
         */
        items: {
            wiredName: allItemsWiredName,
            getter: function () {
                return this._getROProperty(allItemsWiredName, arguments).then(function (res) {
                    return StringUtils.isNullOrEmpty(res) ? [] : res.split("\n");
                });
            }
        }
    },

    /**
     * @ignore
     * @memberof WPF.ComboBoxTOBehavior
     */
    _methods:{
        /*jshint eqnull:true */
        /**
         * Selects an item in the combobox.
         * @function
         * @param {number|string} indexOrItem The item to be selected.
         * The input can be the 0-based index of the item or the text of the item.
         * @returns {Promise<void>} a promise that is fulfilled when the select operation ends.
         */
        select: function (indexOrItem) {
            if (indexOrItem == null)
                throw new Error("The parameter indexOrItem cannot be null.");
            return this._executeWithEvents(arguments, "Select", indexOrItem);
        }
    }
};

module.exports.ComboBox = ComboBox;
module.exports.ComboBoxTO = ComboBoxTO;
