// leanft.sdk.wpf/lib/table_cell.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 10/05/2016.
 */
var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var MergeBehavior = Core.MergeBehavior;

/**
 * A single cell in a WPF Table.
 * @constructor
 * @class TableCell
 * @mixes WPF.TableCellBehavior._properties
 * @mixes WPF.TableCellBehavior._methods
 * @memberof WPF
 */
function TableCell(ownerTable, rowIdentifier, columnIdentifier) {
    Utils.extend(this, MergeBehavior);
    this._mergeBehavior(TableCellBehavior);
    this._ownerTable = ownerTable;
    this._rowIdentifier = rowIdentifier;
    this._columnIdentifier = columnIdentifier;
    this._executeWithEventsOnItem = this._ownerTable._executeWithEventsOnItem.bind(this._ownerTable, this);
    this._executeWithEventsOnItemOnError = this._ownerTable._executeWithEventsOnItemOnError.bind(this._ownerTable, this);
}

TableCell.prototype = {
    _ownerTable: null,
    _rowIdentifier: null,
    _columnIdentifier: null,
    _executeWithEventsOnItem: null,
    _executeWithEventsOnItemOnError: null,
};

/**
 * The methods of TableCell object
 * @memberof WPF
 * @ignore
 */
var TableCellBehavior = {
    /**
     * @memberof WPF.TableCellBehavior
     * @ignore
     */
    _properties: {
        /**
         * The x-coordinate (in pixels) of the cell relative to the table.
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the x-coordinate (in pixels) of the cell relative to the table.
         * @memberof WPF.TableCell#
         */
        x: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellX, this._rowIdentifier, this._columnIdentifier);
        },

        /**
         * The y-coordinate (in pixels) of the cell relative to the table.
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the y-coordinate (in pixels) of the cell relative to the table.
         * @memberof WPF.TableCell#
         */
        y: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellY, this._rowIdentifier, this._columnIdentifier);
        },

        /**
         * The height of the cell (in pixels).
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the height of the cell (in pixels).
         * @memberof WPF.TableCell#
         */
        height: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellHeight, this._rowIdentifier, this._columnIdentifier);
        },

        /**
         * The width of the cell (in pixels).
         * @function
         * @returns {Promise<number>} a promise that is fulfilled with the width of the cell (in pixels).
         * @memberof WPF.TableCell#
         */
        width: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellWidth, this._rowIdentifier, this._columnIdentifier);
        },

        /**
         * The content of the cell.
         * @function
         * @returns {Promise<Object>} a promise that is fulfilled with the content of the cell.
         * @memberof WPF.TableCell#
         */
        value: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellValue, this._rowIdentifier, this._columnIdentifier);
        },

        /**
         * The column header name of the cell.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the column header name of the cell.
         * @memberof WPF.TableCell#
         */
        columnName: function () {
            return this._executeWithEventsOnItemOnError(this._ownerTable._getCellColumnName, this._rowIdentifier, this._columnIdentifier);
        }
    },

    /**
     * @memberof WPF.TableCellBehavior
     * @ignore
     */
    _methods: {
        /**
         * Sets the cell to a new value.
         * @function
         * @param {string|number} value The new value to be set.
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         * @memberof WPF.TableCell#
         */
        setValue: function (value) {
            return this._executeWithEventsOnItem(this._ownerTable._setCellValue, this._rowIdentifier, this._columnIdentifier, value);
        }
    }
};

module.exports.TableCell = TableCell;
