// leanft.sdk.wpf/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var WPFFactory = require("./lib/wpf_factory.js").WPFFactory;
var DescriptionsAndEnums = {
    Window: require("./lib/window.js").Window,    
    UiObject: require("./lib/ui_object.js").UiObject,
    ComboBox: require("./lib/combo_box.js").ComboBox,
    Image: require("./lib/image.js").Image,
    ProgressBar: require("./lib/progress_bar.js").ProgressBar,
    StatusBar: require("./lib/status_bar.js").StatusBar,
    ScrollBar: require("./lib/scroll_bar.js").ScrollBar,
    Link: require("./lib/link.js").Link,
    CheckBox: require("./lib/check_box.js").CheckBox,
    RadioButton: require("./lib/radio_button.js").RadioButton,
    TabStrip: require("./lib/tab_strip.js").TabStrip,
    List: require("./lib/list.js").List,
    Edit: require("./lib/edit.js").Edit,
    Button: require("./lib/button.js").Button,
    TreeView: require("./lib/tree_view.js").TreeView,
    Menu: require("./lib/menu.js").Menu,
    Slider: require("./lib/slider.js").Slider,
    ToolBar: require("./lib/tool_bar.js").ToolBar,
    Calendar: require("./lib/calendar.js").Calendar,
    Table: require("./lib/table.js").Table
};

module.exports.WPFFactory = WPFFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
