// leanft.sdk.winforms/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var WinFormsFactory = require("./lib/winforms_factory.js").WinFormsFactory;
var DescriptionsAndEnums = {
     UiObject: require("./lib/ui_object.js").UiObject,
     Window: require("./lib/window.js").Window,
     ComboBox: require("./lib/combo_box.js").ComboBox,
     Edit: require("./lib/edit.js").Edit,
     Editor: require("./lib/editor.js").Editor,
     Label: require("./lib/label.js").Label,
     Calendar: require("./lib/calendar.js").Calendar,
     CheckBox: require("./lib/check_box.js").CheckBox,
     Button: require("./lib/button.js").Button,
     ListBox: require("./lib/list_box.js").ListBox,
     ScrollBar: require("./lib/scroll_bar.js").ScrollBar,
     SpinBox: require("./lib/spin_box.js").SpinBox,
     TabControl: require("./lib/tab_control.js").TabControl,
     ToolBar: require("./lib/tool_bar.js").ToolBar,
     ListView: require("./lib/list_view.js").ListView,
     PropertyGrid: require("./lib/propertygrid.js").PropertyGrid,
     RadioButton: require("./lib/radio_button.js").RadioButton,
     StatusBar: require("./lib/status_bar.js").StatusBar,
     TreeView: require("./lib/tree_view.js").TreeView,
     Table: require("./lib/table.js").Table    
};

module.exports.WinFormsFactory = WinFormsFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
