// leanft.sdk.java/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..

var JavaFactory = require("./lib/java_factory.js").JavaFactory;
var DescriptionsAndEnums = {
    UiObject: require("./lib/ui_object.js").UiObject,
    Window: require("./lib/window.js").Window,
    Applet: require("./lib/applet.js").Applet,
    Dialog: require("./lib/dialog.js").Dialog,
    InternalFrame: require("./lib/internal_frame.js").InternalFrame,
    Button: require("./lib/button.js").Button,
    Editor: require("./lib/editor.js").Editor,
    Link: require("./lib/link.js").Link,
    TabControl: require("./lib/tab_control.js").TabControl,
    Menu: require("./lib/menu.js").Menu,
    Label: require("./lib/label.js").Label,
    Calendar: require("./lib/calendar.js").Calendar,
    SpinBox: require("./lib/spin_box.js").SpinBox,
    Table: require("./lib/table.js").Table,
    TableKeyModifier: require("./lib/table.js").TableKeyModifier,
    ToolBar: require("./lib/tool_bar.js").ToolBar,
    RadioButton: require("./lib/radio_button.js").RadioButton,
    ExpandBar: require("./lib/expand_bar.js").ExpandBar,
    TreeView: require("./lib/tree_view.js").TreeView,
    TreeViewTable: require("./lib/tree_view_table.js").TreeViewTable,
    CheckBox: require("./lib/check_box.js").CheckBox,
    List: require("./lib/list.js").List,
    Slider: require("./lib/slider.js").Slider,
    SliderOrientation: require("./lib/slider.js").SliderOrientation,
    UiEvent: require("./lib/event_info.js").UiEvent,
    UiEventInfo: require("./lib/event_info.js").UiEventInfo,
    KeyboardEvent: require("./lib/event_info.js").KeyboardEvent,
    KeyboardEventInfo: require("./lib/event_info.js").KeyboardEventInfo,
    MouseEvent: require("./lib/event_info.js").MouseEvent,
    MouseEventInfo: require("./lib/event_info.js").MouseEventInfo
};

module.exports.JavaFactory = JavaFactory;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
