// leanft.sdk.oracle/lib/radio_group.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 09/18/2018.
 */

var Core = require("leanft.sdk.core");
var ObjectUtils = Core.ObjectUtils;
var UiObjectBase = require("./ui_object_base.js");
var HelperBehaviors = require("./helper_behaviors.js");
var ClassPathDescriptionBehavior = HelperBehaviors.ClassPathDescriptionBehavior;
var ClassPathInfoProvider = HelperBehaviors.ClassPathInfoProvider;
var EnabledDescriptionBehavior = HelperBehaviors.EnabledDescriptionBehavior;
var EnabledInfoProvider = HelperBehaviors.EnabledInfoProvider;
var ObjectInfoDescriptionBehavior = HelperBehaviors.ObjectInfoDescriptionBehavior;
var ObjectInfoProvider = HelperBehaviors.ObjectInfoProvider;
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var RadioGroupItem = require("./radio_group_item.js").RadioGroupItem;
var Convert = Core.Convert;
var WrappedItemExecutorBehavior = Core.ExecuteItemsWithEventsBehavior;
var WaitUntilEnabledBehavior = Core.WaitUntilEnabledBehavior;

/**
 * RadioGroupDescriptionBehavior - the description for an Oracle option (radio button) group.
 * @memberof Oracle
 * @ignore
 */
var RadioGroupDescriptionBehavior = {
    /**
     * The value of the selected item.
     * @param {string} [selectedItem] When provided, this method acts as a setter and sets the selectedItem value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored selectedItem value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    selectedItem: function (selectedItem) {
        return this._propertyHandler("selected item", selectedItem, Convert.toString);
    },
    /**
     * The caption of the object.
     * @param {string} [caption] When provided, this method acts as a setter and sets the caption value.
     * @returns {string | Description} If no parameter is provided, the method acts as a getter and returns the current stored caption value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    caption: function (caption) {
        return this._propertyHandler("prompt", caption, Convert.toString);
    },
    /**
     * The 0-based index of the selected item.
     * @param {number} [selectedIndex] When provided, this method acts as a setter and sets the selectedIndex value.
     * NOTE: If no item is selected, -1 is returned.
     * @returns {number | Description} If no parameter is provided, the method acts as a getter and returns the current stored selectedIndex value;
     * otherwise it returns 'this' for use with the builder pattern.
     */
    selectedIndex: function (selectedIndex) {
        return this._propertyHandler("selected index", selectedIndex, function (index) {
            index = Convert.toNumber(index);
            return index != null ? index + 1 : null;
        }, function (index) {
            index = Convert.toNumber(index);
            return index != null ? index - 1 : null;
        });
    }
};

/**
 * The description of an Oracle option (radio button) group.
 * @param {Object} [props] Initializes a RadioGroup test object using the supplied property bag.
 * @returns {RadioGroup} the description object that was initialized.
 * @class RadioGroup
 * @extends Oracle.UiObjectBaseDescription
 * @mixes Oracle.ClassPathDescriptionBehavior
 * @mixes Oracle.EnabledDescriptionBehavior
 * @mixes Oracle.ObjectInfoDescriptionBehavior
 * @mixes Oracle.RadioGroupDescriptionBehavior
 * @memberof Oracle
 */
function RadioGroup(props) {
    var radioGroupDesc = new UiObjectBaseDescription("RadioGroup");
    ObjectUtils.extend(radioGroupDesc, EnabledDescriptionBehavior);
    ObjectUtils.extend(radioGroupDesc, ClassPathDescriptionBehavior);
    ObjectUtils.extend(radioGroupDesc, ObjectInfoDescriptionBehavior);
    ObjectUtils.extend(radioGroupDesc, RadioGroupDescriptionBehavior);
    radioGroupDesc._defineByPropertyBag(props);
    return radioGroupDesc;
}

/**
 * An Oracle option (radio button) group.
 * @constructor
 * @class RadioGroupTO
 * @extends Oracle.UiObjectBaseTO
 * @mixes Oracle.RadioGroupTOBehavior._properties
 * @mixes Oracle.ClassPathInfoProvider._properties
 * @mixes Oracle.EnabledInfoProvider._properties
 * @mixes Oracle.ObjectInfoProvider._properties
 * @mixes Oracle.RadioGroupTOBehavior._methods
 * @mixes WaitUntilEnabledBehavior._methods
 * @memberof Oracle
 */
function RadioGroupTO(session, description) {
    description = description || RadioGroup();
    var radioGroupTO = new UiObjectBaseTO("OracleRadioGroup", session, description);
    radioGroupTO._mergeBehavior(WrappedItemExecutorBehavior);
    radioGroupTO._mergeBehavior(ClassPathInfoProvider);
    radioGroupTO._mergeBehavior(EnabledInfoProvider);
    radioGroupTO._mergeBehavior(ObjectInfoProvider);
    radioGroupTO._mergeBehavior(WaitUntilEnabledBehavior);
    radioGroupTO._mergeBehavior(RadioGroupTOBehavior);
    return radioGroupTO;
}

/**
 * The behaviors of an Oracle option (radio button) group.
 * @memberof Oracle
 * @ignore
 */
var RadioGroupTOBehavior = {

    _helpers: {
        _itemTransform: function (item) {
            if (typeof item === 'object')
                item = item._getIdentifier();

            if (typeof item === 'number')
                return item + 1;

            if (typeof item === 'string')
                return item;

            throw new Error("ArgumentException: unsupported identifier type");
        },

        /**
         * Checks whether the specified item is selected.
         * @function
         * @param {number|string} identifier The identifier of the item. If it is a number, it is the 0-based index of the item.
         * If it is a string, it is the name of the item.
         * @returns {Promise<string>} a promise that is fulfilled with the item's text when the _getItemText operation ends.
         * @ignore
         */
        _isItemSelected: function (identifier) {
            return this._executeNotWrapped(arguments, "IsSelected", this._itemTransform(identifier));
        },

        /**
         * Create an item from the identifier
         * @function
         * @param {number|string} identifier The identifier of the item. If it is a number, it is the 0-based index of the item.
         * @returns {RadioGroupItem} The RadioGroupItem object
         * @ignore
         */
        _createItem: function (identifier) {
            return new RadioGroupItem(this, identifier);
        }
    },
    /**
     * @ignore
     * @memberof Oracle.RadioGroupTOBehavior
     */
    _properties: {
        /**
         * The caption of the object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the caption.
         */
        caption: {
            wiredName: "prompt"
        },
        /**
         * The value of the selected item.
         * @function
         * @returns {Promise<RadioGroupItem>} a promise that is fulfilled with the selectedItem.
         */
        selectedItem: {
            wiredName: "selected item",
            getter: function () {
                return this._getROProperty("selected item", arguments).then((function (item) {
                    return this._createItem(item);
                }).bind(this));
            }
        },
        /**
         * The 0-based index of the selected item.
         * @function
         * NOTE: If no item is selected, -1 is returned.
         * @returns {Promise<number>} a promise that is fulfilled with the selectedIndex.
         */
        selectedIndex: {
            wiredName: "selected index",
            getter: function () {
                return this._getROProperty("selected index", arguments).then(function (index) {
                    return index - 1;
                });
            }
        }
    },

    /**
     * @ignore
     * @memberof Oracle.RadioGroupTOBehavior
     */
    _methods: {
        /*jshint eqnull:true */
        /**
         * Returns the item with the specified text.
         * @function
         * @param {string|number} item The text/index of the item.
         * @memberof Oracle.RadioGroupTO#
         */
        getItem: function (item) {
            if (item == null)
                throw new Error("ArgumentException: text is required.");
            return this._createItem(item);
        },

        /**
         * Transfers the focus to the option group.
         * @function
         * @returns {Promise<void>} a promise that is fulfilled when the setFocus operation ends.
         */
        setFocus: function () {
            return this._executeWithEvents(arguments, "SetFocus");
        },

        /*jshint unused:true, eqnull:true */
        /**
         * Selects the radio button with the specified value.
         * @function
         * @param {RadioGroupItem | number | string} item The item to be selected.
         * The input can be the radio group item, the 0-based index of the item or the name/text of the item
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        select: function (item) {
            if (item == null)
                throw new Error("ArgumentException: items is required.");
            return this._executeWithEvents(arguments, "Select", this._itemTransform(item));
        }
    }
};

module.exports.RadioGroup = RadioGroup;
module.exports.RadioGroupTO = RadioGroupTO;
