// leanft.sdk.sapgui/lib/component_type.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/24/2016.
 */


/**
 * SAPGUI component types.
 * Possible values for the 'ComponentType' type.
 * @typedef {SAPGUI.ComponentType} ComponentType
 * @property {string} unknown Unknown type.
 * @property {string} component Component type.
 * @property {string} vComponent Vcomponent type.
 * @property {string} vContainer Vcontainer type.
 * @property {string} application Application
 * @property {string} connection Connection type.
 * @property {string} session Session type.
 * @property {string} frameWindow Frame Window type.
 * @property {string} mainWindow Main Window type.
 * @property {string} modalWindow Modal Window type.
 * @property {string} messageWindow Message Window type.
 * @property {string} label Label type.
 * @property {string} textField Text Field type.
 * @property {string} cTextField Ctextfield type.
 * @property {string} passwordField Password Field type.
 * @property {string} comboBox Combobox type.
 * @property {string} okCodeField OKCode Field type.
 * @property {string} button Button type.
 * @property {string} radioButton Radio Button type.
 * @property {string} checkBox Checkbox type.
 * @property {string} statusPane Status Pane type.
 * @property {string} customControl Custom Control type.
 * @property {string} containerShell Container Shell type.
 * @property {string} box Box type.
 * @property {string} container Container type.
 * @property {string} simpleContainer Simple Container type.
 * @property {string} scrollContainer Scroll Container type.
 * @property {string} listContainer List Container type.
 * @property {string} userArea User Area type.
 * @property {string} tableControl Table Control type.
 * @property {string} tableColumn tablecolumn type.
 * @property {string} tableRow Tablerow type.
 * @property {string} tabStrip Tabstrip type.
 * @property {string} tab Tab type.
 * @property {string} scrollbar Scrollbar type.
 * @property {string} toolBar Toolbar type.
 * @property {string} titlebar Titlebar type.
 * @property {string} statusbar StatusBar type
 * @property {string} menu Menu type.
 * @property {string} menubar Menubar type.
 * @property {string} collection Collection type.
 * @property {string} sessionInfo Session Info type.
 * @property {string} shell Shell type.
 * @property {string} gosShell GOS Shell type.
 * @property {string} splitterShell Splitter Shell type.
 * @property {string} dialogShell Dialog Shell type.
 * @property {string} dockShell Dock Shell type.
 * @property {string} contextMenu Context Menu type.
 * @property {string} componentCollection Component Collection type.
 * @property {string} ctrlTree TreeView control type.
 * @property {string} ctrlGridView Grid View control type.
 * @property {string} ctrlToolBar toolbar control type.
 * @property {string} ctrlTextArea Editor control type.
 * @property {string} gridToolBar Grid toolbar type.
 * @property {string} ctrlCalendar Calendar control type.
 * @property {string} fallbackComponent Fallback Component type.
 * @property {string} ctrlApoGrid APOGrid control type.
 * @memberof SAPGUI
 */
var ComponentType = {
    /**
     * Unknown type.
     * @type {string}
     */
    unknown: "unknown",

    /**
     * Component type.
     * @type {string}
     */
    component: "component",

    /**
     * vcomponent type.
     * @type {string}
     */
    vComponent: "vcomponent",

    /**
     * vcontainer type.
     * @type {string}
     */
    vContainer: "vcontainer",

    /**
     * Application type.
     * @type {string}
     */
    application: "application",

    /**
     * connection type.
     * @type {string}
     */
    connection: "connection",

    /**
     * Session type.
     * @type {string}
     */
    session: "session",

    /**
     * Frame Window type.
     * @type {string}
     */
    frameWindow: "framewindow",

    /**
     * Main Window type.
     * @type {string}
     */
    mainWindow: "mainwindow",

    /**
     * Modal Window type.
     * @type {string}
     */
    modalWindow: "modalwindow",

    /**
     * Message Window type.
     * @type {string}
     */
    messageWindow: "messagewindow",

    /**
     * Label type.
     * @type {string}
     */
    label: "label",

    /**
     * Text Field type.
     * @type {string}
     */
    textField: "textfield",

    /**
     * Ctextfield type.
     * @type {string}
     */
    cTextField: "ctextfield",

    /**
     * Password Field type.
     * @type {string}
     */
    passwordField: "passwordfield",

    /**
     * combobox type.
     * @type {string}
     */
    comboBox: "combobox",

    /**
     * OKCode Field type.
     * @type {string}
     */
    okCodeField: "okcodefield",

    /**
     * Button type.
     * @type {string}
     */
    button: "button",

    /**
     * Radio Button type.
     * @type {string}
     */
    radioButton: "radiobutton",

    /**
     * checkbox type.
     * @type {string}
     */
    checkBox: "checkbox",

    /**
     * Status Pane type.
     * @type {string}
     */
    statusPane: "statuspane",

    /**
     * Custom Control type.
     * @type {string}
     */
    customControl: "customcontrol",

    /**
     * Container Shell type.
     * @type {string}
     */
    containerShell: "containershell",

    /**
     * Box type.
     * @type {string}
     */
    box: "box",

    /**
     * Container type.
     * @type {string}
     */
    container: "container",

    /**
     * Simple Container type.
     * @type {string}
     */
    simpleContainer: "simplecontainer",

    /**
     * Scroll Container type.
     * @type {string}
     */
    scrollContainer: "scrollcontainer",

    /**
     * List Container type.
     * @type {string}
     */
    listContainer: "listcontainer",

    /**
     * User Area type.
     * @type {string}
     */
    userArea: "userarea",

    /**
     * Table Control type.
     * @type {string}
     */
    tableControl: "tablecontrol",

    /**
     * table column type.
     * @type {string}
     */
    tableColumn: "tablecolumn",

    /**
     * table row type.
     * @type {string}
     */
    tableRow: "tablerow",

    /**
     * tab strip type.
     * @type {string}
     */
    tabStrip: "tabstrip",

    /**
     * Tab type.
     * @type {string}
     */
    tab: "tab",

    /**
     * Scrollbar type.
     * @type {string}
     */
    scrollbar: "scrollbar",

    /**
     * toolbar type.
     * @type {string}
     */
    toolBar: "toolbar",

    /**
     * Titlebar type.
     * @type {string}
     */
    titlebar: "titlebar",

    /**
     * StatusBar type
     * @type {string}
     */
    statusbar: "statusbar",

    /**
     * Menu type.
     * @type {string}
     */
    menu: "menu",

    /**
     * Menubar type.
     * @type {string}
     */
    menubar: "menubar",

    /**
     * Collection type.
     * @type {string}
     */
    collection: "collection",

    /**
     * Session Info type.
     * @type {string}
     */
    sessionInfo: "sessioninfo",

    /**
     * Shell type.
     * @type {string}
     */
    shell: "shell",

    /**
     * GOS Shell type.
     * @type {string}
     */
    gosShell: "gosshell",

    /**
     * Splitter Shell type.
     * @type {string}
     */
    splitterShell: "splittershell",

    /**
     * Dialog Shell type.
     * @type {string}
     */
    dialogShell: "dialogshell",

    /**
     * Dock Shell type.
     * @type {string}
     */
    dockShell: "dockshell",

    /**
     * Context Menu type.
     * @type {string}
     */
    contextMenu: "contextmenu",

    /**
     * Component Collection type.
     * @type {string}
     */
    componentCollection: "componentcollection",

    /**
     * TreeView control type.
     * @type {string}
     */
    ctrlTree: "ctrltree",

    /**
     * Grid View control type.
     * @type {string}
     */
    ctrlGridView: "ctrlgridview",

    /**
     * toolbar control type.
     * @type {string}
     */
    ctrlToolBar: "ctrltoolbar",

    /**
     * Editor control type.
     * @type {string}
     */
    ctrlTextArea: "ctrltextarea",

    /**
     * Grid toolbar type.
     * @type {string}
     */
    gridToolBar: "gridtoolbar",

    /**
     * Calendar control type.
     * @type {string}
     */
    ctrlCalendar: "ctrlcalendar",

    /**
     * Fallback Component type.
     * @type {string}
     */
    fallbackComponent: "fallbackcomponent",

    /**
     * APOGrid control type.
     * @type {string}
     */
    ctrlApoGrid: "ctrlapogrid",

    /**
     * @ignore
     */
    _convertToComponentType: {
        "-1": "unknown",
        "0": "component",
        "1": "vcomponent:",
        "2": "vcontainer",
        "10": "application",
        "11": "connection",
        "12": "session",
        "20": "framewindow",
        "21": "mainwindow",
        "22": "modalwindow",
        "23": "messagewindow",
        "30": "label",
        "31": "textfield",
        "32": "ctextfield",
        "33": "passwordfield",
        "34": "combobox",
        "35": "okcodefield",
        "40": "button",
        "41": "radiobutton",
        "42": "checkbox",
        "43": "statuspane",
        "50": "customcontrol",
        "51": "containershell",
        "62": "box",
        "70": "container",
        "71": "simplecontainer",
        "72": "scrollcontainer",
        "73": "listcontainer",
        "74": "userarea",
        "80": "tablecontrol",
        "81": "tablecolumn",
        "82": "tablerow",
        "90": "tabstrip",
        "91": "tab",
        "100": "scrollbar",
        "101": "toolbar",
        "102": "titlebar",
        "103": "statusbar",
        "110": "menu",
        "111": "menubar",
        "120": "collection",
        "121": "sessioninfo",
        "122": "shell",
        "123": "gosshell",
        "124": "splittershell",
        "125": "dialogshell",
        "126": "dockshell",
        "127": "contextmenu",
        "128": "componentcollection",
        "200": "ctrltree",
        "201": "ctrlgridview",
        "202": "ctrltoolbar",
        "203": "ctrltextarea",
        "204": "gridtoolbar",
        "205": "ctrlcalendar",
        "206": "fallbackcomponent",
        "207": "ctrlapogrid"
    },

    /**
     * @ignore
     */
    _convertToNative: {},

    /**
     * @ignore
     */
    _nativeToComponentType: function (type) {
        return ComponentType._convertToComponentType[type.toString().toLowerCase()];
    },

    /**
     * @ignore
     */
    _componentTypeToNative: function (type) {
        return ComponentType._convertToNative[type.toLowerCase()];
    }
};

Object.keys(ComponentType._convertToComponentType).forEach(function(key) {
    var value = ComponentType._convertToComponentType[key];
    ComponentType._convertToNative[value] = parseInt(key);
});

module.exports.ComponentType = ComponentType;
