// leanft.sdk.sapgui/lib/sap_function_key.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/24/2016.
 */

/**
 * Function keys and key combinations defined by SAPGUI.
 * @typedef {SAPGUI.SapFunctionKey} SapFunctionKey
 * @property {string} enter Enter key.
 * @property {string} f1 F1 key.
 * @property {string} f2 F2 key.
 * @property {string} f3 F3 key.
 * @property {string} f4 F4 key.
 * @property {string} f5 F5 key.
 * @property {string} f6 F6 key.
 * @property {string} f7 F7 key.
 * @property {string} f8 F8 key.
 * @property {string} f9 F9 key.
 * @property {string} f10 F10 key.
 * @property {string} f11 F11 key.
 * @property {string} esc F12 key.
 * @property {string} shiftF1 Shift + F1 key.
 * @property {string} shiftF2 Shift + F2 key.
 * @property {string} shiftF3 Shift + F3 key.
 * @property {string} shiftF4 Shift + F4 key.
 * @property {string} shiftF5 Shift + F5 key.
 * @property {string} shiftF6 Shift + F6 key.
 * @property {string} shiftF7 Shift + F7 key.
 * @property {string} shiftF8 Shift + F8 key.
 * @property {string} shiftF9 Shift + F9 key.
 * @property {string} shiftF10 Shift + F10 key.
 * @property {string} shiftF11 Shift + F11 key.
 * @property {string} shiftF12 Shift + F12 key.
 * @property {string} ctrlF1 Ctrl + F1 key.
 * @property {string} ctrlF2 Ctrl + F2 key.
 * @property {string} ctrlF3 Ctrl + F3 key.
 * @property {string} ctrlF4 Ctrl + F4 key.
 * @property {string} ctrlF5 Ctrl + F5 key.
 * @property {string} ctrlF6 Ctrl + F6 key.
 * @property {string} ctrlF7 Ctrl + F7 key.
 * @property {string} ctrlF8 Ctrl + F8 key.
 * @property {string} ctrlF9 Ctrl + F9 key.
 * @property {string} ctrlF10 Ctrl + F10 key.
 * @property {string} ctrlF11 Ctrl + F11 key.
 * @property {string} ctrlF12 Ctrl + F12 key.
 * @property {string} ctrlShiftF1 Ctrl + Shift + F1.
 * @property {string} ctrlShiftF2 Ctrl + Shift + F2.
 * @property {string} ctrlShiftF3 Ctrl + Shift + F3.
 * @property {string} ctrlShiftF4 Ctrl + Shift + F4.
 * @property {string} ctrlShiftF5 Ctrl + Shift + F5.
 * @property {string} ctrlShiftF6 Ctrl + Shift + F6.
 * @property {string} ctrlShiftF7 Ctrl + Shift + F7.
 * @property {string} ctrlShiftF8 Ctrl + Shift + F8.
 * @property {string} ctrlShiftF9 Ctrl + Shift + F9.
 * @property {string} ctrlShiftF10 Ctrl + Shift + F10.
 * @property {string} ctrlShiftF11 Ctrl + Shift + F11.
 * @property {string} ctrlShiftF12 Ctrl + Shift + F12.
 * @property {string} pageDown Page Down key.
 * @property {string} pageUp Page Up key.
 * @property {string} ctrlF Ctrl + F keys.
 * @property {string} ctrlPageUp Ctrl + PageUp keys.
 * @property {string} ctrlPageDown Ctrl + PageDown keys.
 * @property {string} ctrlG Ctrl + G keys.
 * @property {string} ctrlP Ctrl + P keys.
 * @memberof SAPGUI
 */
var SapFunctionKey = {
    /**
     * Enter key.
     * @type {string}
     */
    enter: "enter",

    /**
     * F1 key.
     * @type {string}
     */
    f1: "f1",

    /**
     * F2 key.
     * @type {string}
     */
    f2: "f2",

    /**
     * F3 key.
     * @type {string}
     */
    f3: "f3",

    /**
     * F4 key.
     * @type {string}
     */
    f4: "f4",

    /**
     * F5 key.
     * @type {string}
     */
    f5: "f5",

    /**
     * F6 key.
     * @type {string}
     */
    f6: "f6",

    /**
     * F7 key.
     * @type {string}
     */
    f7: "f7",

    /**
     * F8 key.
     * @type {string}
     */
    f8: "f8",

    /**
     * F9 key.
     * @type {string}
     */
    f9: "f9",

    /**
     * F10 key.
     * @type {string}
     */
    f10: "f10",

    /**
     * F11 key.
     * @type {string}
     */
    f11: "f11",

    /**
     * F12 key.
     * @type {string}
     */
    esc: "esc",

    /**
     * Shift + F1 key.
     * @type {string}
     */
    shiftF1: "shiftf1",

    /**
     * Shift + F2 key.
     * @type {string}
     */
    shiftF2: "shiftf2",

    /**
     * Shift + F3 key.
     * @type {string}
     */
    shiftF3: "shiftf3",

    /**
     * Shift + F4 key.
     * @type {string}
     */
    shiftF4: "shiftf4",

    /**
     * Shift + F5 key.
     * @type {string}
     */
    shiftF5: "shiftf5",

    /**
     * Shift + F6 key.
     * @type {string}
     */
    shiftF6: "shiftf6",

    /**
     * Shift + F7 key.
     * @type {string}
     */
    shiftF7: "shiftf7",

    /**
     * Shift + F8 key.
     * @type {string}
     */
    shiftF8: "shiftf8",

    /**
     * Shift + F9 key.
     * @type {string}
     */
    shiftF9: "shiftf9",

    /**
     * Shift + F10 key.
     * @type {string}
     */
    shiftF10: "shiftf10",

    /**
     * Shift + F11 key.
     * @type {string}
     */
    shiftF11: "shiftf11",

    /**
     * Shift + F12 key.
     * @type {string}
     */
    shiftF12: "shiftf12",

    /**
     *  Ctrl + F1 key.
     * @type {string}
     */
    ctrlF1: "ctrlf1",

    /**
     *  Ctrl + F2 key.
     * @type {string}
     */
    ctrlF2: "ctrlf2",

    /**
     *  Ctrl + F3 key.
     * @type {string}
     */
    ctrlF3: "ctrlf3",

    /**
     *  Ctrl + F4 key.
     * @type {string}
     */
    ctrlF4: "ctrlf4",

    /**
     *  Ctrl + F5 key.
     * @type {string}
     */
    ctrlF5: "ctrlf5",

    /**
     *  Ctrl + F6 key.
     * @type {string}
     */
    ctrlF6: "ctrlf6",

    /**
     *  Ctrl + F7 key.
     * @type {string}
     */
    ctrlF7: "ctrlf7",

    /**
     *  Ctrl + F8 key.
     * @type {string}
     */
    ctrlF8: "ctrlf8",

    /**
     *  Ctrl + F9 key.
     * @type {string}
     */
    ctrlF9: "ctrlf9",

    /**
     *  Ctrl + F10 key.
     * @type {string}
     */
    ctrlF10: "ctrlf10",

    /**
     *  Ctrl + F11 key.
     * @type {string}
     */
    ctrlF11: "ctrlf11",

    /**
     *  Ctrl + F12 key.
     * @type {string}
     */
    ctrlF12: "ctrlf12",

    /**
     * Ctrl + Shift + F1.
     * @type {string}
     */
    ctrlShiftF1: "ctrlshiftf1",

    /**
     * Ctrl + Shift + F2.
     * @type {string}
     */
    ctrlShiftF2: "ctrlshiftf2",

    /**
     * Ctrl + Shift + F3.
     * @type {string}
     */
    ctrlShiftF3: "ctrlshiftf3",

    /**
     * Ctrl + Shift + F4.
     * @type {string}
     */
    ctrlShiftF4: "ctrlshiftf4",

    /**
     * Ctrl + Shift + F5.
     * @type {string}
     */
    ctrlShiftF5: "ctrlshiftf5",

    /**
     * Ctrl + Shift + F6.
     * @type {string}
     */
    ctrlShiftF6: "ctrlshiftf6",

    /**
     * Ctrl + Shift + F7.
     * @type {string}
     */
    ctrlShiftF7: "ctrlshiftf7",

    /**
     * Ctrl + Shift + F8.
     * @type {string}
     */
    ctrlShiftF8: "ctrlshiftf8",

    /**
     * Ctrl + Shift + F9.
     * @type {string}
     */
    ctrlShiftF9: "ctrlshiftf9",

    /**
     * Ctrl + Shift + F10.
     * @type {string}
     */
    ctrlShiftF10: "ctrlshiftf10",

    /**
     * Ctrl + Shift + F11.
     * @type {string}
     */
    ctrlShiftF11: "ctrlshiftf11",

    /**
     * Ctrl + Shift + F12.
     * @type {string}
     * */
    ctrlShiftF12: "ctrlshiftf12",


    /**
     * Page Down key.
     * @type {string}
     */
    pageDown: "pagedown",

    /**
     * Page Up key.
     * @type {string}
     */
    pageUp: "pageup",

    /**
     * Ctrl + F keys.
     * @type {string}
     */
    ctrlF: "ctrlf",

    /**
     * Ctrl + PageUp keys.
     * @type {string}
     */
    ctrlPageUp: "ctrlpageup",

    /**
     * Ctrl + PageDown keys.
     * @type {string}
     */
    ctrlPageDown: "ctrlpagedown",

    /**
     * Ctrl + G keys.
     * @type {string}
     */
    ctrlG: "ctrlg",

    /**
     * Ctrl + P keys.
     * @type {string}
     */
    ctrlP: "ctrlp",

    /**
     * @ignore
     */
    _convertToNative: {
        "enter": 0,
        "f1": 1,
        "f2": 2,
        "f3": 3,
        "f4": 4,
        "f5": 5,
        "f6": 6,
        "f7": 7,
        "f8": 8,
        "f9": 9,
        "f10": 10,
        "f11": 11,
        "esc": 12,
        "shiftf1": 13,
        "shiftf2": 14,
        "shiftf3": 15,
        "shiftf4": 16,
        "shiftf5": 17,
        "shiftf6": 18,
        "shiftf7": 19,
        "shiftf8": 20,
        "shiftf9": 21,
        "shiftf10": 22,
        "shiftf11": 23,
        "shiftf12": 24,
        "ctrlf1": 25,
        "ctrlf2": 26,
        "ctrlf3": 27,
        "ctrlf4": 28,
        "ctrlf5": 29,
        "ctrlf6": 30,
        "ctrlf7": 31,
        "ctrlf8": 32,
        "ctrlf9": 33,
        "ctrlf10": 34,
        "ctrlf11": 35,
        "ctrlf12": 36,
        "ctrlshiftf1": 37,
        "ctrlshiftf2": 38,
        "ctrlshiftf3": 39,
        "ctrlshiftf4": 40,
        "ctrlshiftf5": 41,
        "ctrlshiftf6": 42,
        "ctrlshiftf7": 43,
        "ctrlshiftf8": 44,
        "ctrlshiftf9": 45,
        "ctrlshiftf10": 46,
        "ctrlshiftf11": 47,
        "ctrlshiftf12": 48,
        "pagedown": 82,
        "pageup": 81,
        "ctrlf": 71,
        "ctrlpageup": 80,
        "ctrlpagedown": 83,
        "ctrlg": 84,
        "ctrlp": 86
    },

    /**
     * @ignore
     */
    _sapFunctionKeyToNative: function (type) {
        return SapFunctionKey._convertToNative[type.toLowerCase()];
    }

};

module.exports.SapFunctionKey = SapFunctionKey;
