// leanft.sdk.core/lib/window_state_and_styles.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 06/02/2016.
 */

/**
 * Represents the window's current state (maximized/minimized).
 * @enum
 */
var WindowState = {
    /**
     * The window is maximized.
    * @type {string}
    */
    maximized: "Maximized",
    /**
     * The window is minimized.
    * @type {string}
    */
    minimized: "Minimized",
    /**
     * The window is in normal-restored state.
    * @type {string}
    */
    normal: "Normal"
};

/**
 * The Window Styles enum.
 * @enum
 */
var WindowStyles = {
    /**
    * The window has a thin-line border.
    * @type {number}
    */
    ws_BORDER: 0x00800000,

    /**
    *The window has a title bar (includes the ws_BORDER style).
    * @type {number}
    */
    ws_CAPTION: 0x00C00000,

    /**
    *The window is a child window. 
    *A window with this style cannot have a menu bar. 
    *This style cannot be used with the ws_POPUP style.
    * @type {number}
    */
    ws_CHILD: 0x40000000,

    /**
    *Same as the ws_CHILD style.
    * @type {number}
    */
    ws_CHILDWINDOW: 0x40000000,

    /**
    *Excludes the area occupied by child windows when drawing occurs within the parent window. 
    *This style is used when creating the parent window.
    * @type {number}
    */
    ws_CLIPCHILDREN: 0x02000000,

    /**
    *Clips child windows relative to each other. That is, when a particular child window receives a WM_PAINT message,
    *the ws_CLIPSIBLINGS style clips all other overlapping child windows out of the region of the child window to be updated. 
    *If ws_CLIPSIBLINGS is not specified and child windows overlap, it is possible,
    *when drawing within the client area of a child window, to draw within the client area of a neighboring child window.
    * @type {number}
    */
    ws_CLIPSIBLINGS: 0x04000000,

    /**
    *The window is initially disabled. A disabled window cannot receive input from the user.      
    * @type {number}
    */
    ws_DISABLED: 0x08000000,

    /**
    *The window has a border of a style typically used with dialog boxes. 
    *A window with this style cannot have a title bar.
    * @type {number}
    */
    ws_DLGFRAME: 0x00400000,

    /**
    *The window is the first control of a group of controls. 
    *The group consists of this first control and all controls defined after it, up to the next control with the ws_GROUP style. 
    * @type {number}
    */
    ws_GROUP: 0x00020000,

    /**
    *The window has a horizontal scroll bar.
    * @type {number}
    */
    ws_HSCROLL: 0x00100000,

    /**
    *The window is initially minimized. Same as the ws_MINIMIZE style.
    * @type {number}
    */
    ws_ICONIC: 0x20000000,

    /**
    *The window is initially maximized.
    * @type {number}
    */
    ws_MAXIMIZE: 0x01000000,

    /**
    *The window has a maximize button. 
    * @type {number}
    */
    ws_MAXIMIZEBOX: 0x00010000,

    /**
    *The window is initially minimized. Same as the ws_ICONIC style.
    * @type {number}
    */
    ws_MINIMIZE: 0x20000000,

    /**
    *The window has a minimize button.       
    * @type {number}
    */
    ws_MINIMIZEBOX: 0x00020000,

    /**
    *The window is an overlapped window. An overlapped window has a title bar and a border. 
    *Same as the ws_TILED style.
    * @type {number}
    */
    ws_OVERLAPPED: 0x00000000,

    /**
    *The window is an overlapped window. 
    *Same as the ws_TILEDWINDOW style.
    * @type {number}
    */
    ws_OVERLAPPEDWINDOW: null,

    /**
    *The window is a pop-up window.
    * @type {number}
    */
    ws_POPUP: 0x80000000,

    /**
    *The window is a pop-up window. 
    * @type {number}
    */
    ws_POPUPWINDOW: null,

    /**
    *The window has a sizing border. 
    *Same as the ws_THICKFRAME style.
    * @type {number}
    */
    ws_SIZEBOX: 0x00040000,

    /**
    *The window has a window menu on its title bar. 
    * @type {number}
    */
    ws_SYSMENU: 0x00080000,

    /**
    *The window is a control that can receive the keyboard focus when the user presses the TAB key.  
    * @type {number}
    */
    ws_TABSTOP: 0x00010000,

    /**
    *The window has a sizing border. 
    *Same as the ws_SIZEBOX style.
    * @type {number}
    */
    ws_THICKFRAME: 0x00040000,

    /**
    *The window is an overlapped window. An overlapped window has a title bar and a border. 
    *Same as the ws_OVERLAPPED style.
    * @type {number}
    */
    ws_TILED: 0x00000000,

    /**
    *The window is an overlapped window. 
    *Same as the ws_OVERLAPPEDWINDOW style.
    * @type {number}
    */
    ws_TILEDWINDOW: null,

    /**
    *The window is initially visible.
    * @type {number}
    */
    ws_VISIBLE: 0x10000000,

    /**
    *The window has a vertical scroll bar.
    * @type {number}
    */
    ws_VSCROLL: 0x00200000,
};

WindowStyles.ws_OVERLAPPEDWINDOW = WindowStyles.ws_OVERLAPPED | WindowStyles.ws_CAPTION | WindowStyles.ws_SYSMENU | WindowStyles.ws_THICKFRAME | WindowStyles.ws_MINIMIZEBOX | WindowStyles.ws_MAXIMIZEBOX;
WindowStyles.ws_POPUPWINDOW = WindowStyles.ws_POPUP | WindowStyles.ws_BORDER | WindowStyles.ws_SYSMENU;
WindowStyles.ws_TILEDWINDOW = WindowStyles.ws_OVERLAPPED | WindowStyles.ws_CAPTION | WindowStyles.ws_SYSMENU | WindowStyles.ws_THICKFRAME | WindowStyles.ws_MINIMIZEBOX | WindowStyles.ws_MAXIMIZEBOX;
    

/**
 * Represents extended window style properties.
 * @enum
 */
var WindowExtendedStyles = {
    /**
    *The window accepts drag and drop files.
    * @type {number}
    */
    ws_EX_ACCEPTFILES: 0x00000010,

    /**
    *Forces a top-level window onto the taskbar when the window is visible.
    * @type {number}
    */
    ws_EX_APPWINDOW: 0x00040000,

    /**
    *The window has a border with a sunken edge.
    * @type {number}
    */
    ws_EX_CLIENTEDGE: 0x00000200,

    /**
    *Paints all descendants of a window in bottom-to-top painting order using double-buffering.       
    * @type {number}
    */
    ws_EX_COMPOSITED: 0x02000000,

    /**
    *The title bar of the window includes a question mark. When the user clicks the question mark, the cursor changes to a question mark with a pointer.
    * @type {number}
    */
    ws_EX_CONTEXTHELP: 0x00000400,

    /**
    *The window contains child windows that take part in dialog box navigation.
    * @type {number}
    */
    ws_EX_CONTROLPARENT: 0x00010000,

    /**
    *The window has a double border.
    * @type {number}
    */
    ws_EX_DLGMODALFRAME: 0x00000001,

    /**
    *The window is a layered window.
    * @type {number}
    */
    ws_EX_LAYERED: 0x00080000,

    /**
    *If the shell language is Hebrew, Arabic, or another language that supports reading order alignment, 
    *the horizontal origin of the window is on the right edge. 
    *Increasing horizontal values advance to the left.
    *For other languages, the style is ignored.
    * @type {number}
    */
    ws_EX_LAYOUTRTL: 0x00400000,

    /**
    *The window has generic left-aligned properties. 
    *This is the default.
    * @type {number}
    */
    ws_EX_LEFT: 0x00000000,

    /**
    *If the shell language is Hebrew, Arabic, or another language that supports reading order alignment, 
    *the vertical scroll bar (if present) is to the left of the client area. 
    *For other languages, the style is ignored.
    * @type {number}
    */
    ws_EX_LEFTSCROLLBAR: 0x00004000,

    /**
    *The window text is displayed using left-to-right reading-order properties. 
    *This is the default.
    * @type {number}
    */
    ws_EX_LTRREADING: 0x00000000,

    /**
    *The window is an MDI child window.
    * @type {number}
    */
    ws_EX_MDICHILD: 0x00000040,

    /**
    *A top-level window created with this style does not become the foreground window when the user clicks it.
    * @type {number}
    */
    ws_EX_NOACTIVATE: 0x08000000,

    /**
    *The window does not pass its window layout to its child windows.
    * @type {number}
    */
    ws_EX_NOINHERITLAYOUT: 0x00100000,

    /**
    *A child window created with this style does not send the WM_PARENTNOTIFY message to its parent window 
    *when it is created or destroyed.
    * @type {number}
    */
    ws_EX_NOPARENTNOTIFY: 0x00000004,

    /**
    *The window does not render to a redirection surface. 
    *This is for windows that do not have visible content or that use mechanisms other than surfaces to provide their visual.
    * @type {number}
    */
    ws_EX_NOREDIRECTIONBITMAP: 0x00200000,

    /**
    *The window is an overlapped window.
    * @type {number}
    */
    ws_EX_OVERLAPPEDWINDOW: null,

    /**
    *The window is a palette window, which is a modeless dialog box that presents an array of commands.
    * @type {number}
    */
    ws_EX_PALETTEWINDOW: null,

    /**
    *The window has generic right-aligned properties.
    * @type {number}
    */
    ws_EX_RIGHT: 0x00001000,

    /**
    *The vertical scroll bar (if present) is to the right of the client area. 
    *This is the default.
    * @type {number}
    */
    ws_EX_RIGHTSCROLLBAR: 0x00000000,

    /**
    *If the shell language is Hebrew, Arabic, or another language that supports reading-order alignment, 
    *the window text is displayed using right-to-left reading-order properties. 
    *For other languages, the style is ignored.
    * @type {number}
    */
    ws_EX_RTLREADING: 0x00002000,

    /**
    *The window has a three-dimensional border style intended to be used for items that do not accept user input.
    * @type {number}
    */
    ws_EX_STATICEDGE: 0x00020000,

    /**
    *The window is intended to be used as a floating toolbar.
    * @type {number}
    */
    ws_EX_TOOLWINDOW: 0x00000080,

    /**
    *The window is placed above all non-topmost windows and stays above them, 
    *even when the window is deactivated.
    * @type {number}
    */
    ws_EX_TOPMOST: 0x00000008,

    /**
    *The window is not painted until siblings beneath the window 
    *(that were created by the same thread) have been painted.
    * @type {number}
    */
    ws_EX_TRANSPARENT: 0x00000020,

    /**
    *The window has a border with a raised edge.
    * @type {number}
    */
    ws_EX_WINDOWEDGE: 0x00000100,
};

WindowExtendedStyles.ws_EX_OVERLAPPEDWINDOW = WindowExtendedStyles.ws_EX_WINDOWEDGE | WindowExtendedStyles.ws_EX_CLIENTEDGE;
WindowExtendedStyles.ws_EX_PALETTEWINDOW = WindowExtendedStyles.ws_EX_WINDOWEDGE | WindowExtendedStyles.ws_EX_TOOLWINDOW | WindowExtendedStyles.ws_EX_TOPMOST;

module.exports.WindowState = WindowState;
module.exports.WindowStyles = WindowStyles;
module.exports.WindowExtendedStyles = WindowExtendedStyles;
