// leanft.report/lib/const.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
var Path = require("path");

var Const = {
        DEFAULT_TITLE:"ReplayObjectNotFoundException",
        DEFAULT_DESCRIPTION:"ReplayObjectNotUniqueException",
        DEFAULT_REPORT_FOLDER_NAME:"ReplayParentMissingException",
        ENV_VAR_LEANFT_REPORTBUILDER_DIR:"ReplayObjectNotVisibleException",
        ENV_VAR_LEANFT_REPORTS_TARGETDIR:"ReplayObjectDisabledException",
        ENV_VAR_COMPUTERNAME:"ReplayItemNotFoundException",
        ENV_VAR_NUMBER_OF_PROCESSORS:"ReplayItemNotUniqueException",
        HTML_REPORT_BUILDER_EXE_NAME:"ReplayTimeoutException",
        Root : "Results",
        ReportNode : "ReportNode",
        GeneralInfo : "GeneralInfo",
        Data : "Data",
        ToolName : "ToolName",
        CustomFields : "CustomFields",
        Environment : "Environment",
        TestedApplication : "TestedApplication",
        Extension : "Extension",
        Any : "Any",
        IconKey : "IconKey",
        Properties : "Properties",
        Property : "Property",
        TestObject : "TestObject",
        Object : "Object",
        Path : "Path",
        HideContextInfo : "HideContextInfo",
        ApplicationModel : "ApplicationModel",
        Descriptive : "Descriptive",
        LogicalName : "LogicalName",
        ClassName : "ClassName",
        CustomUIElements : "CustomUIElements",
        DescriptionLabel : "DescriptionLabel",
        FlowTreeNodeCaption : "FlowTreeNodeCaption",
        FlowTreeNodeTooltip : "FlowTreeNodeTooltip",
        Verification : "Verification",
        ErrorListItemStatusIndicatorIconKey : "ErrorListItemStatusIndicatorIconKey",
        HideFlowTreeNodeStatusIndicator : "HideFlowTreeNodeStatusIndicator",
        IsHtmlErrorText : "IsHtmlErrorText",
        Name : "Name",
        Description : "Description",
        Type : "Type",
        Value : "Value",
        StartTime : "StartTime",
        Duration : "Duration",
        Result : "Result",
        ErrorText : "ErrorText",
        Links : "Links",
        Snapshot : "Snapshot",
        StepCustomImagePath : "StepCustomImagePath",
        Repository : "Repository",
        Operation : "Operation",
        OperationData : "OperationData",
        RunStartTime : "RunStartTime",
        Timezone : "Timezone",
        XmlLocale : "Locale",
        OrchestrationToolName : "OrchestrationToolName",
        OrchestrationToolVersion : "OrchestrationToolVersion",
        ReportTitle : "ReportTitle",
        Field : "Field",
        OSInfo : "OSInfo",
        HostName : "HostName",
        NumberOfCores : "NumberOfCores",
        XmlCpuInfo : "CpuInfo",
        XmlTotalMemory : "TotalMemory",
        XmlUser : "User",
        RuntimeEngineInfo : "RuntimeEngineInfo",
        Version : "Version",
        ReportFileName : "runresults",
        DefaultTitle : "Run Results",
        HtmlTableCssStyle :  "\x3Cstyle\x3E.ex-table{border-collapse: collapse !important;width: auto;height: auto;}.ex-table,.ex-table tr,.ex-table td{border: 1px solid #c6c6c6;}.ex-table td{padding: 5px 10px;vertical-align: top,width: auto,}.ex-table .ex-name{font-size: 13px;font-weight: bold;background-color: #f3f3f3;white-space: nowrap;}.ex-table .ex-value{color: #5a5a5a;}\x3C\x2Fstyle\x3E",
        VriPropertyName : "VRI",
        VriPropertyValue : "*Test object description contains also VRI definitions",
        Title : "Title",
        TitleToolTip : "TitleToolTip",
        ArgumentsTitle : "Data",
        CSS : "CSS",
        XPath : "XPath",
        MobileServerAddress : "Mobile Center Server:",
        EnvVarLeanftBPTTest : "LEANFT_BPT_RUN_DATA",
        EnvVarLeanftTestDataFields : "LEANFT_TEST_DATA_FIELDS",
        EnvironmentFile : "EnvData.xml",
        OperationDataTitle : "OperationDataTitle",
        MachineName: "machine_name",
        OsType: "os_type",
        OsVersion: "os_version",
        RunTime: "run_time",
        CpuInfo: "cpu_info",
        NumberOfCpu: "number_of_cpu",
        TotalMemory: "TotalMemory",
        TimeZone: "time_zone",
        Locale: "Locale",
        User: "User",
        SnapshotsFolder : Path.join("Resources","Snapshots"),
        UserFolder : Path.join("Resources","User"),
        CData : "#cdata-section",
};


module.exports.Const = Const;
