// leanft.report/lib/empty_reporter_behavior.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
EmptyReporterBehavior = {
    cleanup: function () { },

    generateReport: function () { },

    startScreenRecording: function (useCompression, outputPath) { },

    stopScreenRecording: function () { },

    reportEvent: function (stepName, description, status) { },

    reportEventWithImage: function (stepName, description, status, image) { },

    reportEventWithAdditionalInfo: function (stepName, additionalInfo) {},

    reportErrorEvent: function (stepName, description, exception) { },

    reportErrorEventWithImage: function (stepName, description, error, image) { },

    reportVerification: function (status, verificationData) { },

    addRunInformation: function (fieldName, fieldValue) { },

    startReportingContext: function (contextName, contextDescription, additionalData) { },

    startReportingContextWithInfo: function (contextName, reportEventInfo) { },

    endReportingContext: function () { },

    setSnapshotCaptureLevel: function (snapshotCaptureLevel) { },

    startSpec: function (specName, specDescription) { },

    endSpec: function (status) { },

    reportExpectEvent: function (data) { },

    getCurrentExecutionContextStatus: function(){ }
};

module.exports = EmptyReporterBehavior;
