// leanft.sdk.web/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
module.exports.BrowserFactory = require("./lib/browser_factory.js").BrowserFactory;
module.exports.BrowserType = require("./lib/browser_factory.js").BrowserType;
module.exports.WebFactory = require("./lib/web_factory.js").WebFactory;
module.exports.WebElement = require("./lib/webelement.js").WebElement;
module.exports.Event = require("./lib/webelement.js").Event;
module.exports.MouseEvent = require("./lib/webelement.js").MouseEvent;
module.exports.EmulatedDevice = require("./lib/emulated_device.js").EmulatedDevice;

//descriptions
module.exports.Link = require("./lib/link.js").Link;
module.exports.Button = require("./lib/button.js").Button;
module.exports.Edit = require("./lib/edit.js").Edit;
module.exports.RadioGroup = require("./lib/radio_group.js").RadioGroup;
module.exports.ListBox = require("./lib/list_box.js").ListBox;
module.exports.CheckBox = require("./lib/check_box.js").CheckBox;
module.exports.Image = require("./lib/image.js").Image;
module.exports.ImageTypes = require("./lib/image.js").ImageTypes;
module.exports.DialogButton = require("./lib/browser.js").DialogButton;
module.exports.AgGrid = require("./lib/aggrid.js").AgGrid;
module.exports.UiGrid = require("./lib/uigrid.js").UiGrid;
module.exports.Area = require("./lib/area.js").Area;
module.exports.Page = require("./lib/page.js").Page;
module.exports.Frame = require("./lib/frame.js").Frame;
module.exports.AudioControl = require("./lib/audio_control.js").AudioControl;
module.exports.VideoControl = require("./lib/video_control.js").VideoControl;
module.exports.FileField = require("./lib/file_field.js").FileField;
module.exports.Table = require("./lib/table.js").Table;
module.exports.Menu = require("./lib/menu.js").Menu;
module.exports.TabStrip = require("./lib/tab_strip.js").TabStrip;
module.exports.TreeView = require("./lib/tree.js").TreeView;
module.exports.NumericField = require("./lib/numeric_field.js").NumericField;
module.exports.Slider = require("./lib/slider.js").Slider;

//Internal Behaviors to be used by SAPUI5
module.exports.Behaviors = require("./lib/web_behaviors.js");
module.exports.HelperFuncs = require("./lib/helper_behaviors.js").HelperFuncs;

//Export Test Object needed for SAPUI5 unit tests
module.exports.EditTO = require("./lib/edit.js").EditTO;
module.exports.ButtonTO = require("./lib/button.js").ButtonTO;
module.exports.TableTO = require("./lib/table.js").TableTO;
module.exports.CheckBoxTO = require("./lib/check_box.js").CheckBoxTO;
