// leanft.sdk.web/lib/emulated_device.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created  on 12/04/2016.
 */

/**
 * Emulated Device class containing predefined emulated devices.
 * @ignore
 */
var EmulatedDevice = {

    /**
     * Predefined Samsung Galaxy S5 device.
     */
    GalaxyS5: {
        name: "Samsung Galaxy S5",
        width: 360,
        height: 640,
        scaleFactor: 3,
        userAgent: "Mozilla/5.0 (Linux; Android 5.0; SAMSUNG SM-G900F/G900FXXU1POK5 Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/2.1 Chrome/34.0.1847.76 Mobile Safari/537.36",
        isTouchable: true,
        isMobile: true
    },

    /**
     * Predefined Google Nexus 5X device.
     */
    Nexus5X: {
        name: "Google Nexus 5X",
        width: 411,
        height: 731,
        scaleFactor: 2.627,
        userAgent: "Mozilla/5.0 (Linux; Android 6.0; Nexus 5X Build/MDB08I) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.43 Mobile Safari/537.36",
        isTouchable: true,
        isMobile: true
    },

    /**
     * Predefined Google Nexus 6P device.
     */
    Nexus6P: {
        name: "Google Nexus 6P",
        width: 435,
        height: 773,
        scaleFactor: 3.311,
        userAgent: "Mozilla/5.0 (Linux; Android 6.0.1; Nexus 6P Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.83 Mobile Safari/537.36",
        isTouchable: true,
        isMobile: true
    },

    /**
     * Predefined Apple iPhone 5 device.
     */
    IPhone5: {
        name: "Apple iPhone 5",
        width: 320,
        height: 568,
        scaleFactor: 2,
        userAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X; en-us) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53",
        isTouchable: true,
        isMobile: true
    },

    /**
     * Predefined Apple iPhone 6 device.
     */
    IPhone6: {
        name: "Apple iPhone 6",
        width: 375,
        height: 667,
        scaleFactor: 2,
        userAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 8_0 like Mac OS X) AppleWebKit/600.1.3 (KHTML, like Gecko) Version/8.0 Mobile/12A4345d Safari/600.1.4",
        isTouchable: true,
        isMobile: true
    },

    /**
     * Predefined Apple iPhone 6 Plus device.
     */
    IPhone6Plus: {
        name: "Apple iPhone 6 Plus",
        width: 414,
        height: 736,
        scaleFactor: 3,
        userAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 8_0 like Mac OS X) AppleWebKit/600.1.3 (KHTML, like Gecko) Version/8.0 Mobile/12A4345d Safari/600.1.4",
        isTouchable: true,
        isMobile: true
    },

    IPad: {
        name: "Apple iPad",
        width: 768,
        height: 1024,
        scaleFactor: 2,
        userAgent: "Mozilla/5.0 (iPad; CPU OS 7_0 like Mac OS X) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53",
        isTouchable: true,
        isMobile: true
    },
};

var _deviceMap = {
    "galaxy s5" : EmulatedDevice.GalaxyS5,
    "nexus 5x" : EmulatedDevice.Nexus5X,
    "nexus 6p" : EmulatedDevice.Nexus6P,
    "iphone 5" : EmulatedDevice.IPhone5,
    "iphone 6" : EmulatedDevice.IPhone6,
    "iphone 6 plus" : EmulatedDevice.IPhone6Plus,
    "ipad" : EmulatedDevice.IPad,
};

/*jshint eqnull:true */
var EmulatedDeviceConverter = {

    convert: function(emulatedDevice){
        var emulatedDeviceObj = _deviceMap[emulatedDevice.toLowerCase()];
        if (emulatedDeviceObj == null) {
            throw new Error("Predefined emulated device '" + emulatedDevice + "' not found.");
        }
        return emulatedDeviceObj;
    },
};

module.exports.EmulatedDevice = EmulatedDevice;
module.exports.EmulatedDeviceConverter = EmulatedDeviceConverter;
