// leanft.sdk.web/lib/uigrid.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 19/08/2018.
 * This file contains the definition of the UiGrid test object and the description.
 */

var ObjUtils = require("leanft.sdk.core").ObjectUtils;
var ClickBehavior = require("leanft.sdk.core").ClickBehavior;
var DragAndDropBehavior = require("leanft.sdk.core").DragAndDropBehavior;
var ExecuteItemsWithEventsBehavior = require("leanft.sdk.core").ExecuteItemsWithEventsBehavior;
var WebBase = require("./web_testobject_base.js").WebBaseTestObject;
var WebBaseDesc = require("./web_testobject_base.js").WebBaseDescription;
var WebElement = require("./webelement.js");
var TableTestObjectBehavior = require("./table.js").TableTestObjectBehavior;
var GridBaseDescriptionBehavior = require("./grid_base.js").GridBaseDescriptionBehavior;
var GridBaseTOBehavior = require("./grid_base.js").GridBaseTOBehavior;
var Convert = require("leanft.sdk.core").Convert;


/**
 * The description of a Web UiGrid test object.
 * @param {Object} [props] Initializes the description using the supplied property bag.
 * @returns {UiGrid} the description object that was initialized.
 * @constructor
 * @memberof Web
 * @extends Web.WebBaseDescription
 * @mixes Web.WebElementDescriptionBehaviour
 * @mixes Web.GridBaseDescriptionBehavior
 */
function UiGrid(props) {

    var tableDes = new WebBaseDesc("UiGrid"); // leanFTType

    ObjUtils.extend(tableDes, WebElement.WebElementDescriptionBehaviour);
    ObjUtils.extend(tableDes, GridBaseDescriptionBehavior);
    tableDes._defineByPropertyBag(props);

    return tableDes;
}

/**
 * A Web UiGrid containing a variable number of rows and columns.
 * @class UiGridTO
 * @memberof Web
 * @constructor
 * @extends Web.WebBaseTestObject
 * @mixes Web.WebElementTOBehavior._properties
 * @mixes Web.WebElementTOBehavior._methods
 * @mixes ClickBehavior._methods
 * @mixes DragAndDropBehavior._methods
 * @mixes Web.TableTestObjectBehavior._properties
 * @mixes Web.TableTestObjectBehavior._methods
 * @mixes Web.GridBaseTOBehavior._properties
 * @mixes Web.GridBaseTOBehavior._methods
 * @mixes Web.UiGridTestObjectBehavior._methods
 */
function UiGridTO(session, description) {
    description = description || UiGrid();
    var tableTO = new WebBase("WebUIGrid", session, description); // native class

    tableTO._mergeBehavior(ClickBehavior);
    tableTO._mergeBehavior(DragAndDropBehavior);
    tableTO._mergeBehavior(WebElement.WebElementTOBehavior);
    tableTO._mergeBehavior(TableTestObjectBehavior);
    tableTO._mergeBehavior(GridBaseTOBehavior);
    tableTO._mergeBehavior(UiGridTestObjectBehavior);
    tableTO._mergeBehavior(ExecuteItemsWithEventsBehavior);

    return tableTO;
}

//Consts
var findRowWiredName = "GetRowWithCellText";

/**
 * @memberof Web
 * @ignore
 */
var UiGridTestObjectBehavior = {
    /**
     * @memberof Web.UiGridTestObjectBehavior
     * @ignore
     */
    _methods: {
        /*jshint eqnull:true */
        /**
         * Scrolls to a specified row, as defined by the row’s number.
         * @function
         * @param {number} row 0-based row number
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        scrollToRow: function (row) {
            if (row == null)
                throw new Error("The parameter row cannot be null.");
            return this._executeWithEvents(arguments, "ScrollToRow", Convert.toNumber(row) + 1);
        },

        /*jshint eqnull:true */
        /**
         * Finds the first row with the specified text in the specified column, starting from the specified row.
         * @param {string} cellText Text to find.
         * @param {number|string} [column] The 0-based column index or column header.
         * @param {number} [startFromRow = 0] The index from which to start the search.
         * @returns {Promise<TableCell[]>} a promise that is fulfilled either with an array of cells for the requested row, or fulfilled with null if no match was found.
         * @function
         */
        findRow: {
            wiredName: findRowWiredName,
            exec: function (cellText, column, startFromRow) {
                return this._findRow(cellText, column, startFromRow);
            }
        }
    }
};

module.exports.UiGrid = UiGrid;
module.exports.UiGridTO = UiGridTO;
