// leanft.sdk.web/lib/web_behaviors.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 12/28/2015.
 */

module.exports.HelperFuncs = require("./helper_behaviors.js").HelperFuncs;
module.exports.BuildPathBehavior = require("./helper_behaviors.js").BuildPathBehavior;

// Descriptions
module.exports.WebBaseDescription = require("./web_testobject_base.js").WebBaseDescription;
module.exports.WebElementDescriptionBehavior = require("./webelement.js").WebElementDescriptionBehaviour;
module.exports.IsEnabledBehavior = require("./helper_behaviors.js").IsEnabledBehavior;
module.exports.ButtonDescriptionBehavior = require("./button.js").ButtonDescriptionBehavior;
module.exports.FormFieldBehavior = require("./helper_behaviors.js").FormFieldBehavior;
module.exports.CheckBoxDescriptionBehavior = require("./check_box.js").CheckBoxDescriptionBehavior;
module.exports.EditFieldBaseDescriptionBehavior = require("./edit_field_base.js").EditFieldBaseDescriptionBehavior;
module.exports.EditDescriptionBehavior = require("./edit.js").EditDescriptionBehavior;
module.exports.ListBoxDescriptionBehavior = require("./list_box.js").ListBoxDescriptionBehavior;
module.exports.TableDescriptionBehavior = require("./table.js").TableDescriptionBehavior;


// Test Objects
module.exports.WebBaseTestObject = require("./web_testobject_base.js").WebBaseTestObject;
module.exports.WebElementTOBehavior = require("./webelement.js").WebElementTOBehavior;
module.exports.ButtonTOBehavior = require("./button.js").ButtonTOBehavior;
module.exports.CheckBoxTOBehavior = require("./check_box.js").CheckBoxTOBehavior;
module.exports.EditFieldBaseTestObjectBehavior = require("./edit_field_base.js").EditFieldBaseTestObjectBehavior;
module.exports.EditTOBehavior = require("./edit.js").EditTOBehavior;
module.exports.ListBoxTOBehavior = require("./list_box.js").ListBoxTOBehavior;
module.exports.TableTestObjectBehavior = require("./table.js").TableTestObjectBehavior;
module.exports.TableFindRowTOBehavior = require("./table.js").TableFindRowTOBehavior;
module.exports.TableBorderTOBehavior = require("./table.js").TableBorderTOBehavior;
module.exports.TableCellBehavior = require("./table.js").TableCell;
