// leanft.sdk.mobile/lib/stepper.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 12/22/2015.
 */

var ObjUtils = require("leanft.sdk.core").ObjectUtils;
var UiObjectBase = require("./uiobject_base").UiObjectBaseTestObject;
var UiObjectBaseDesc = require("./uiobject_base").UiObjectBaseDescription;


/**
 * The description of a Stepper test object in a mobile application.
 * @param {Object} [props] Initializes the description using the supplied property bag.
 * @returns {Stepper} the description object that was initialized.
 * @constructor
 * @memberof Mobile
 * @extends Mobile.UiObjectBaseDescription
 */
function Stepper(props) {
    var stepperDes = new UiObjectBaseDesc("Stepper");

    stepperDes._defineByPropertyBag(props);

    return stepperDes;
}

/**
 * A stepper object in a mobile application. (iOS Only)
 * @class StepperTO
 * @memberof Mobile
 * @constructor
 * @extends Mobile.UiObjectBaseTestObject
 * @mixes Mobile.StepperTOBehavior._methods
 */
function StepperTO(session, description) {
    description = description || Stepper();
    var stepperTO = new UiObjectBase("MobileStepper", session, description);

    stepperTO._mergeBehavior(StepperTOBehavior);

    return stepperTO;
}

//Consts
var upWiredName = "Up";
var downWiredName = "Down";

/**
 * @memberof Mobile
 * @ignore
 */
var StepperTOBehavior = {
    /**
     * @ignore
     * @memberof Mobile.StepperTOBehavior
     */
    _methods: {
        /**
         * Increases the value in the stepper.
         * @function
         * @returns {Promise<void>} a promise that is fulfilled after the value in the stepper increased.
         */
        up: {
            wiredName: upWiredName,
            exec: function () {
                return this._executeWithEvents(arguments, upWiredName);
            }
        },
        /**
         * Decreases the value in the stepper.
         * @function
         * @returns {Promise<void>} a promise that is fulfilled after the value in the stepper decreased.
         */
        down: {
            wiredName: downWiredName,
            exec: function () {
                return this._executeWithEvents(arguments, downWiredName);
            }
        }
    }
};

module.exports.Stepper = Stepper;
module.exports.StepperTO = StepperTO;
