// leanft.sdk.te/lib/emulator_status.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 8/16/2016.
 */

/**
 * Terminal Emulator statuses.
 * @typedef {Object} EmulatorStatus
 * @property {string} ready Successful.
 * @property {string} disconnected Not Connected To Presentation Space.
 * @property {string} busy Presentation Space is Busy.
 * @property {string} locked Inhibited/Keyboard Locked.
 * @property {string} unavailable Unavailable Operation. When working with No HLLAPI.
 * @memberof TerminalEmulators
 */
var EmulatorStatus = {
    /**
     * Successful.
     */
    ready: "ready",
    /**
     * Not connected to presentation space.
     */
    disconnected: "disconnected",
    /**
     * Presentation space is busy.
     */
    busy: "busy",
    /**
     * Inhibited/Keyboard locked.
     */
    locked: "locked",
    /**
     * Unavailable operation. When working with no HLLAPI.
     */
    unavailable: "unavailable"
};

/**
 * @ignore
 * @memberof TerminalEmulators
 */
var EmulatorStatusConverter = {
    /*jshint eqnull:true */
    /**
     * @private
     * @ignore
     * @memberof TerminalEmulators
     */
    _emulatorStatusToNative: function (value) {
        var native = EmulatorStatus[value.toLowerCase()];
        if(native == null) {
            throw new Error("You must supply a valid emulator status");
        }

        return native.charAt(0).toUpperCase() + native.slice(1);
    },
    /**
     * @private
     * @ignore
     * @memberof TerminalEmulators
     */
    _nativeToEmulatorStatus: function (value) {

        var leanftString = EmulatorStatus[value.toLowerCase()];
        if(leanftString == null) {
            throw new Error("You must supply a valid emulator status");
        }

        return leanftString;
    }
};

module.exports.EmulatorStatus = EmulatorStatus;
module.exports.EmulatorStatusConverter = EmulatorStatusConverter;
