// leanft.sdk.te/lib/keys.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 8/19/2016.
 */

/**
 * The Keys enum used in the send key function.
 * @memberof TerminalEmulators
 * @enum
 */
var Keys = {
    /**
     * The PA1 key.
     * @type {string}
     */
    pa1 : "@x",
    /**
     * The PA2 key.
     * @type {string}
     */
    pa2 : "@y",
    /**
     * The PA3 key.
     * @type {string}
     */
    pa3 : "@z",
    /**
     * The PA4 key.
     * @type {string}
     */
    pa4 : "@+",
    /**
     * The PA5 key.
     * @type {string}
     */
    pa5 : "@%",
    /**
     * The PA6 key.
     * @type {string}
     */
    pa6 : "@&",
    /**
     * The PA7 key.
     * @type {string}
     */
    pa7 : "@'",
    /**
     * The PA8 key.
     * @type {string}
     */
    pa8 : "@(",
    /**
     * The PA9 key.
     * @type {string}
     */
    pa9 : "@)",
    /**
     * The PA10 key.
     * @type {string}
     */
    pa10 : "@*",
    /**
     * The PF1 key.
     * @type {string}
     */
    pf1 : "@",
    /**
     * The PF2 key.
     * @type {string}
     */
    pf2 : "@2",
    /**
     * The PF3 key.
     * @type {string}
     */
    pf3 : "@3",
    /**
     * The PF4 key.
     * @type {string}
     */
    pf4 : "@4",
    /**
     * The PF5 key.
     * @type {string}
     */
    pf5 : "@5",
    /**
     * The PF6 key.
     * @type {string}
     */
    pf6 : "@6",
    /**
     * The PF7 key.
     * @type {string}
     */
    pf7 : "@7",
    /**
     * The PF8 key.
     * @type {string}
     */
    pf8 : "@8",
    /**
     * The PF9 key.
     * @type {string}
     */
    pf9 : "@9",
    /**
     * The PF10 key.
     * @type {string}
     */
    pf10 : "@a",
    /**
     * The PF11 key.
     * @type {string}
     */
    pf11 : "@b",
    /**
     * The PF12 key.
     * @type {string}
     */
    pf12 : "@c",
    /**
     * The PF13 key.
     * @type {string}
     */
    pf13 : "@d",
    /**
     * The PF14 key.
     * @type {string}
     */
    pf14 : "@e",
    /**
     * The PF15 key.
     * @type {string}
     */
    pf15 : "@f",
    /**
     * The PF16 key.
     * @type {string}
     */
    pf16 : "@g",
    /**
     * The PF17 key.
     * @type {string}
     */
    pf17 : "@h",
    /**
     * The PF18 key.
     * @type {string}
     */
    pf18 : "@i",
    /**
     * The PF19 key.
     * @type {string}
     */
    pf19 : "@j",
    /**
     * The PF20 key.
     * @type {string}
     */
    pf20 : "@k",
    /**
     * The PF21 key.
     * @type {string}
     */
    pf21 : "@l",
    /**
     * The PF22 key.
     * @type {string}
     */
    pf22 : "@m",
    /**
     * The PF23 key.
     * @type {string}
     */
    pf23 : "@n",
    /**
     * The PF24 key.
     * @type {string}
     */
    pf24 : "@o",
    /**
     * The ALT key.
     * @type {string}
     */
    alt : "@A",
    /**
     * The ALTERNATE_CURSOR key.
     * @type {string}
     */
    alternate_cursor : "@$",
    /**
     * The ATTENTION key.
     * @type {string}
     */
    attention : "@A@Q",
    /**
     * The BACKSPACE key.
     * @type {string}
     */
    backspace : "@<",
    /**
     * The BACKTAB key.
     * @type {string}
     */
    backtab : "@B",
    /**
     * The CLEAR key.
     * @type {string}
     */
    clear : "@C",
    /**
     * The CMD key.
     * @type {string}
     */
    cmd : "@A@Y",
    /**
     * The DOWN key.
     * @type {string}
     */
    down : "@V",
    /**
     * The LEFT key.
     * @type {string}
     */
    left : "@L",
    /**
     * The RIGHT key.
     * @type {string}
     */
    right : "@Z",
    /**
     * The CURSOR_SELECT key.
     * @type {string}
     */
    cursor_select : "@A@J",
    /**
     * The UP key.
     * @type {string}
     */
    up : "@U",
    /**
     * The DELETE key.
     * @type {string}
     */
    delete : "@D",
    /**
     * The DUP key.
     * @type {string}
     */
    dup : "@S@X",
    /**
     * The END key.
     * @type {string}
     */
    end : "@q",
    /**
     * The ENTER key.
     * @type {string}
     */
    enter : "@E",
    /**
     * The ERASE_EOF key.
     * @type {string}
     */
    erase_eof : "@F",
    /**
     * The ERASE_INPUT key.
     * @type {string}
     */
    erase_input : "@A@F",
    /**
     * The FIELD_EXIT key.
     * @type {string}
     */
    field_exit : "@A@E",
    /**
     * The FIELD_MARK key.
     * @type {string}
     */
    field_mark : "@S@y",
    /**
     * The FIELD_MINUS key.
     * @type {string}
     */
    field_minus : "@A@-",
    /**
     * The FIELD_PLUS key.
     * @type {string}
     */
    field_plus : "@A@+",
    /**
     * The HELP key.
     * @type {string}
     */
    help : "@H",
    /**
     * The HEXA key.
     * @type {string}
     */
    hexa : "@A@X",
    /**
     * The HOME key.
     * @type {string}
     */
    home : "@0",
    /**
     * The INSERT key.
     * @type {string}
     */
    insert : "@I",
    /**
     * The INSERT_TOG key.
     * @type {string}
     */
    insert_tog : "@A@I",
    /**
     * The LOCAL_PRINT key.
     * @type {string}
     */
    local_print : "@P",
    /**
     * The NEW_LINE key.
     * @type {string}
     */
    new_line : "@N",
    /**
     * The PAGE_UP key.
     * @type {string}
     */
    page_up : "@u",
    /**
     * The PAGE_DOWN key.
     * @type {string}
     */
    page_down : "@v",
    /**
     * The PRINT key.
     * @type {string}
     */
    print : "@A@t",
    /**
     * The REC_BACKSPACE key.
     * @type {string}
     */
    rec_backspace : "@A@<",
    /**
     * The RESET key.
     * @type {string}
     */
    reset : "@R",
    /**
     * The TAB key.
     * @type {string}
     */
    tab : "@T",
    /**
     * The SHIFT key.
     * @type {string}
     */
    shift : "@S",
    /**
     * The SYS_REQ key.
     * @type {string}
     */
    sys_req : "@A@H",
    /**
     * The TEST key.
     * @type {string}
     */
    test : "@A@c"
};

module.exports.Keys = Keys;
