// leanft.sdk.mobile/app.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..

var MobileFactory = require("./lib/mobile_factory.js").MobileFactory;
var DescriptionsAndEnums = {
    //Enums
    DeviceOrientation: require("./lib/device.js").DeviceOrientation,
    Keys: require("./lib/device.js").Keys,
    //Descriptions
    Device: require("./lib/device.js").Device,
    Application: require("./lib/application.js").Application,
    SimulateAuthFailReason: require("./lib/application.js").SimulateAuthFailReason,
    SimulateAuthCancelOrigin: require("./lib/application.js").SimulateAuthCancelOrigin,
    UiObject: require("./lib/uiobject.js").UiObject,
    Button: require("./lib/button.js").Button,
    Calendar: require("./lib/calendar.js").Calendar,
    CheckBox: require("./lib/check_box.js").CheckBox,
    Label: require("./lib/label.js").Label,
    Edit: require("./lib/edit_field.js").Edit,
    RadioGroup: require("./lib/radio_group.js").RadioGroup,
    Slider: require("./lib/slider.js").Slider,
    Stepper: require("./lib/stepper.js").Stepper,
    Paginator: require("./lib/paginator.js").Paginator,
    DropDown: require("./lib/drop_down.js").DropDown,
    Table: require("./lib/table.js").Table,
    TabStrip: require("./lib/tab_strip.js").TabStrip,
    Toggle: require("./lib/toggle.js").Toggle,
    View: require("./lib/view.js").View,
    WebView: require("./lib/web_view.js").WebView,
    RecyclerView: require("./lib/recycler_view.js").RecyclerView
};

module.exports.MobileFactory = MobileFactory;
module.exports.MobileLab = require("./lib/mobile_lab.js").MobileLab;
module.exports.DescriptionsAndEnums = DescriptionsAndEnums;
