// leanft.sdk.winforms/lib/list_box.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 08/25/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var UiObjectBase = require("./ui_object_base.js");
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var StringUtils = Core.StringUtils;
var ListBaseBehavior = Core.ListBaseBehavior;
var ListItem = require("./list_item.js").ListItem;
var ListBoxDescriptionBaseBehavior = Core.ListBoxDescriptionBaseBehavior;
var ContainerBehavior = Core.ContainerBehavior;
var Convert = Core.Convert;
var ArrayUtils = Core.ArrayUtils;
var WrappedItemExecutorBehavior = Core.ExecuteItemsWithEventsBehavior;


/**
 * The description of a ListBox test object.
 * @param {Object} [props] Initializes a ListBox test object using the supplied property bag.
 * @returns {ListBox} the description object that was initialized.
 * @class ListBox
 * @memberof WinForms
 * @extends WinForms.UiObjectBaseDescription
 * @mixes ListBoxDescriptionBaseBehavior
 */
function ListBox(props) {
    var listBoxDesc = new UiObjectBaseDescription("ListBox");
    Utils.extend(listBoxDesc, ListBoxDescriptionBaseBehavior);
    listBoxDesc._defineByPropertyBag(props);
    return listBoxDesc;
}

/**
 * A ListBox object,
 * @constructor
 * @class ListBoxTO
 * @extends WinForms.UiObjectBaseTO
 * @mixes Core.ContainerBehavior
 * @mixes ListBaseBehavior._methods
 * @mixes Core.WrappedItemExecutorBehavior
 * @mixes WinForms.ListBoxTOBehavior._methods
 * @mixes WinForms.ListBoxTOBehavior._properties
 * @memberof WinForms
 */
function ListBoxTO(session, description) {
    description = description || ListBox();
    var listBoxTO = new UiObjectBaseTO("SwfList", session, description);
    listBoxTO._mergeBehavior(ContainerBehavior);
    listBoxTO._mergeBehavior(ListBaseBehavior);
    listBoxTO._mergeBehavior(WrappedItemExecutorBehavior);
    listBoxTO._mergeBehavior(ListBoxTOBehavior);
    return listBoxTO;
}

/**
 * The behaviors of the ListBox object
 * @memberof WinForms
 * @ignore
 */
var ListBoxTOBehavior = {
    /**
     * @ignore
     * @memberof WinForms.ListBoxTOBehavior
     */
    _methods: {
        /**
         * Activate a list item.
         * @function
         * @param {ListItem | number | string} item The item to be activated.
         * The input can be the list item, the 0-based index of the item or the name/text of the item
         * @param {MouseButton} [button = MouseButton.left] The button to activate. Default is the left button.
         * @returns {Promise<void>} a promise that is fulfilled when the activate operation ends.
         */
        activateItem: function (item, button) {
            return this._activateItem(item, button);
        },

        /**
         * Deselects a list of items.
         * @function
         * @param {...(ListItem | number | string)} items You can pass as many items as like, each item can be either the {@link WinForms.ListItem}, name of the item you wish to select,
         * or the numeric index of the item in the list.
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        deselect: function (items) {
            return this._deselect.apply(this, arguments);
        },

        /**
         * Selects additional items from the multi-selection list.
         * @function
         * @param {...(ListItem | string | number)} items One or more items to add to the selection. Each item can be either a {@link WinForms.ListItem}, the item's name,
         * or the item's numeric index in the list.
         * @returns {Promise<void>} A promise that is fulfilled when the operation is performed.
         */
        extendSelect: function (items) {
            return this._extendSelect.apply(this, arguments);
        },

        /**
         * Selects a list of items.
         * @function
         * @param {...(ListItem | number | string)} items You can pass as many items as like, each item can be either the {@link WinForms.ListItem}, name of the item you wish to select,
         * or the numeric index of the item in the list.
         * @returns {Promise<void>} a promise that is fulfilled when the operation is performed.
         */
        select: function (items) {
            return this._select.apply(this, arguments);
        },

        /**
         * Create a ListItem from its text representation.
         * @function
         * @param {string} text The text representation if the item.
         * @returns {ListItem} The ListItem object. See {@link WinForms.ListItem}
         */
        getItem: function (text) {
            return this._getItem(text);
        }
    },

    /**
     * @ignore
     * @memberof WinForms.ListBoxTOBehavior
     */
    _properties:{
        /**
         * Returns an array of all items in the list box.
         * @function
         * @returns {Promise<ListItem[]>} a promise that is fulfilled with an array of all list items in the listbox. See {@link WinForms.ListItem}
         */
        items: {
            wiredName: ContainerBehavior._allItemsDefaultWiredName,
            getter: function () {
                return this._items("\n");
            }
        },

        /**
         * Returns an array of all selected items in the list box.
         * @function
         * @returns {Promise<ListItem[]>} a promise that is fulfilled with an array of all selected items in the listbox. See {@link WinForms.ListItem}
         */
        selectedItems: {
            wiredName: ContainerBehavior._selectedItemsDefaultWiredName,
            getter: function () {
                return this._selectedItems("\n");
            }
        },

        /**
         * Returns the tooltip text of the error icon associated with this object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the error text of Calendar.
         */
        errorText: function () {
            return this._executeWithEventsOnError(arguments, "GetErrorProviderText");
        }
    },

    _helpers:{
        /**
         * @ignore
         */
        _getItemHeight: function (identifier) {
            return this._executeNotWrapped(arguments, "GetItemProperty", identifier, "height");
        },

        /**
         * @ignore
         */
        _isItemAnchor: function (identifier) {
            return this._executeWithEvents(arguments, "GetItemProperty", identifier, "IsAnchor");
        },

        /**
         * @ignore
         */
        _isItemChecked: function (identifier) {
            return this._executeWithEvents(arguments, "GetItemProperty", identifier, "checked");
        },

        /**
         * Create an item from the identifier and textRepresentation
         * @param {number|string} identifier The identifier of the item. If it is a number, it is the 0-based index of the item.
         * @param {string} textRepresentation The text representation if the item.
         * @returns {ListItem} The ListItem object
         * @ignore
         */
        _createItem: function (identifier, textRepresentation) {
            return new ListItem(this, identifier, textRepresentation);
        }
    },

    _additionalProperties: {
        itemsCount: {wiredName: ContainerBehavior._itemsCountDefaultWiredName}
    }
};

module.exports.ListBox = ListBox;
module.exports.ListBoxTO = ListBoxTO;
