// leanft.sdk.winforms/lib/tree_view.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 09/30/2016.
 */

var Core = require("leanft.sdk.core");
var Utils = Core.ObjectUtils;
var UiObjectBase = require("./ui_object_base.js");
var UiObjectBaseTO = UiObjectBase.UiObjectBaseTO;
var UiObjectBaseDescription = UiObjectBase.UiObjectBaseDescription;
var StringUtils = Core.StringUtils;
var TreeViewDescriptionBaseBehavior = Core.TreeViewDescriptionBaseBehavior;
var TreeViewTOBaseBehavior = Core.TreeViewTOBaseBehavior;
var TreeViewNode = require("./treeview_node.js").TreeViewNode;
var ContainerBehavior = Core.ContainerBehavior;
var MouseButton = Core.MouseButton;
var Convert = Core.Convert;
var WrappedItemExecutorBehavior = Core.ExecuteItemsWithEventsBehavior;

/**
 * The description of a TreeView test object.
 * @param {Object} [props] Initializes a TreeView test object using the supplied property bag.
 * @returns {TreeView} the description object that was initialized.
 * @class TreeView
 * @memberof WinForms
 * @extends WinForms.UiObjectBaseDescription
 * @mixes Core.TreeViewDescriptionBaseBehavior
 */
function TreeView(props) {
    var treeViewDesc = new UiObjectBaseDescription("TreeView");
    Utils.extend(treeViewDesc, TreeViewDescriptionBaseBehavior);
    treeViewDesc._defineByPropertyBag(props);
    return treeViewDesc;
}

/**
 * A TreeView object,
 * @constructor
 * @class TreeViewTO
 * @extends WinForms.UiObjectBaseTO
 * @mixes Core.ContainerBehavior
 * @extends Core.WrappedItemExecutorBehavior
 * @mixes Core.TreeViewTOBaseBehavior
 * @mixes WinForms.TreeViewTOBehavior._methods
 * @mixes WinForms.TreeViewTOBehavior._properties
 * @memberof WinForms
 */
function TreeViewTO(session, description) {
    description = description || TreeView();
    var treeViewTO = new UiObjectBaseTO("SwfTreeView", session, description);
    treeViewTO._mergeBehavior(ContainerBehavior);
    treeViewTO._mergeBehavior(WrappedItemExecutorBehavior);
    treeViewTO._mergeBehavior(TreeViewTOBaseBehavior);
    treeViewTO._mergeBehavior(TreeViewTOBehavior);
    return treeViewTO;
}

/**
 * The behaviors of the TreeView object
 * @memberof WinForms
 * @ignore
 */
var TreeViewTOBehavior = {
    /**
     * @ignore
     * @memberof WinForms.TreeViewTOBehavior
     */
    _methods: {
        /*jshint eqnull:true */
        /**
         * Activate a node.
         * @function
         * @param {TreeViewNode|number|string} node The item to be activated.
         * The input can be the TreeViewNode, the 0-based index of the node or the name/text of the node
         * @param {MouseButton} [button = MouseButton.left] The button to activate. Default is the left button.
         * @returns {Promise<void>} a promise that is fulfilled when the activate operation ends.
         */
        activateNode: function (node, button) {
            return this._activateWithButton(node, button);
        },

        /*jshint eqnull:true */
        /**
         * Select a node.
         * @function
         * @param {TreeViewNode|number|string} node The item to be selected.
         * The input can be the TreeViewNode, the 0-based index of the node or the name/text of the node
         * @param {MouseButton} [button = MouseButton.left] The button to activate. Default is the left button.
         * @returns {Promise<void>} a promise that is fulfilled when the activate operation ends.
         */
        select: function (node, button) {
            return this._selectWithButton(node, button);
        },

        /*jshint eqnull:true */
        /**
         * Create a TreeViewNode from its given path. When the path is not unique, returns the first matching node.
         * @function
         * @param {string} nodePath The path of the node.
         * @returns {TreeViewNode} TreeViewNode object.
         */
        getNode: function (nodePath) {
            if (nodePath == null)
                throw new Error("The parameter text cannot be null.");
            return this._createItem(nodePath, nodePath);
        }
    },

    /**
     * @ignore
     * @memberof WinForms.TreeViewTOBehavior
     */
    _properties:{
        /**
         * Returns an array of all visible nodes in the treeview.
         * @function
         * @returns {Promise<TreeViewNode[]>} a promise that is fulfilled with an array of all visible nodes in the tree view.
         */
        visibleNodes: {
            wiredName: ContainerBehavior._allItemsDefaultWiredName,
            getter: function () {
                return this._items("\n");
            }
        },

        /**
         * Returns an array of all selected nodes in the treeview.
         * @function
         * @returns {Promise<TreeViewNode[]>} a promise that is fulfilled with an array of all selected nodes in the treeview.
         */
        selectedNodes: {
            wiredName: ContainerBehavior._selectedItemsDefaultWiredName,
            getter: function () {
                return this._selectedItems("\n");
            }
        },

        /**
         * Returns an array of all checked nodes in the treeview.
         * @function
         * @returns {Promise<TreeViewNode[]>} a promise that is fulfilled with an array of all checked nodes in the treeview.
         */
        checkedVisibleNodes: function () {
            return this._checkedItems(";", "\n");
        },

        /**
         * Returns the tooltip text of the error icon associated with this object.
         * @function
         * @returns {Promise<string>} a promise that is fulfilled with the error text of TreeView.
         */
        errorText: function () {
            return this._executeWithEventsOnError(arguments, "GetErrorProviderText");
        }
    },

    _helpers: {
        /**
         * Create an node from the identifier and textRepresentation
         * @param {number|string} identifier The identifier of the node. If it is a number, it is the 0-based index of the node.
         * @param {string} textRepresentation The text representation if the node.
         * @returns {TreeViewNode} The TreeViewNode object
         * @ignore
         */
        _createItem: function (identifier, textRepresentation) {
            return new TreeViewNode(this, identifier, textRepresentation);
        }
    },

    _additionalProperties: {
        nodesCount: {wiredName: ContainerBehavior._itemsCountDefaultWiredName}
    }
};

module.exports.TreeView = TreeView;
module.exports.TreeViewTO = TreeViewTO;
