// leanft.sdk.sapui5/lib/navigation_bar.js
// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..

var Core = require("leanft.sdk.core");
var ObjectUtils = Core.ObjectUtils;
var SAPUI5BaseDesc = require("./sapui5_base.js").SAPUI5BaseDesc;
var SAPUI5BaseTestObject = require("./sapui5_base.js").SAPUI5BaseTestObject;
var HelperBehaviors = require("./helper_behaviors.js");

/**
 * The description of an SAPUI5 NavigationBar test object.
 * @param {Object} [props] Initializes a NavigationBar test object using the supplied property bag.
 * @returns {NavigationBar} the description object that was initialized.
 * @constructor
 * @memberof SAPUI5
 * @extends SAPUI5.SAPUI5BaseDesc
 * @mixes SAPUI5.ItemsCountDescriptionBehavior
 * @mixes SAPUI5.ItemsDescriptionBehavior
 */
function NavigationBar(props) {
    var navigationBarDes = new SAPUI5BaseDesc("NavigationBar");
    ObjectUtils.extend(navigationBarDes, HelperBehaviors.ItemsCountDescriptionBehavior);
    ObjectUtils.extend(navigationBarDes, HelperBehaviors.ItemsDescriptionBehavior);

    navigationBarDes._defineByPropertyBag(props);

    return navigationBarDes;
}

/**
 * An SAPUI5 NavigationBar.
 * @constructor
 * @memberof SAPUI5
 * @class NavigationBarTO
 * @extends SAPUI5.SAPUI5BaseTestObject
 * @mixes SAPUI5.ItemsTOBehavior._properties
 * @mixes SAPUI5.SelectedItemTOBehavior._properties
 * @mixes SAPUI5.NavigationBarTOBehavior._methods
 */
function NavigationBarTO(session, description) {
    description = description || NavigationBar();
    var navigationBarTO = new SAPUI5BaseTestObject("SAPUINavigationBar", session, description);
    navigationBarTO._mergeBehavior(HelperBehaviors.ItemsTOBehavior);
    navigationBarTO._mergeBehavior(HelperBehaviors.SelectedItemTOBehavior);
    navigationBarTO._mergeBehavior(NavigationBarTOBehavior);
    return navigationBarTO;
}

//Consts
var openMenuWiredName = "OpenMenu";
var selectWiredName = "Select";

/**
 * @ignore
 * @memberof SAPUI5
 */
var NavigationBarTOBehavior = {
    /**
     * @ignore
     * @memberof SAPUI5.NavigationBarTOBehavior
     */
    _methods: {
        /**
         * Opens the popup menu to show hidden navigation items.
         * @function
         * @returns {Promise<void>} a promise that is fulfilled after the popup menu is opened.
         */
        openMenu: {
            wiredName: openMenuWiredName,
            exec: function () {
                return this._executeWithEvents(arguments, openMenuWiredName);
            }
        },

        /**
         * Selects the specified item (by name) in the navigation bar.
         * @function
         * @param {string} itemText Text of the item to select.
         * @returns {Promise<void>} a promise that is fulfilled after the specified item is selected.
         */
        select: {
            wiredName: selectWiredName,
            exec: function (itemText) {
                if (arguments.length === 0)
                    throw new Error("You must pass one argument to select() method");
                return this._executeWithEvents(arguments, selectWiredName, itemText);
            }
        }
    }
};

module.exports.NavigationBar = NavigationBar;
module.exports.NavigationBarTO = NavigationBarTO;
